// @comicrelief/data-models/schemas/mailer/mocks/requests/MailerRequest.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAILER_REQUESTS = exports.MAILER_REQUESTS_INVALID = exports.MAILER_REQUESTS_VALID = exports.MAILER_REQUEST_VALUES = void 0;
var utils_1 = require("../../../../utils");
var SupporterInfo_mock_1 = require("../../../shared/mocks/SupporterInfo.mock");
exports.MAILER_REQUEST_VALUES = {
    invalid: {
        emailAddress: 'string',
        templateName: null,
        messageParams: true,
    },
    valid: {
        emailAddress: SupporterInfo_mock_1.EMAILS.valid,
        templateName: 'prize_platform__participants',
        messageParams: {
            first_name: 'first_name',
            prize_name: 'prize_name',
            prize_url: 'prize_url',
        },
    },
};
exports.MAILER_REQUESTS_VALID = {
    DEFAULT: exports.MAILER_REQUEST_VALUES.valid,
};
exports.MAILER_REQUESTS_INVALID = {
    DEFAULT: exports.MAILER_REQUEST_VALUES.invalid,
    INVALID_EMAIL_ADDRESS: utils_1.override(exports.MAILER_REQUEST_VALUES.valid, { emailAddress: exports.MAILER_REQUEST_VALUES.invalid.emailAddress }),
    INVALID_TEMPLATE_NAME: utils_1.override(exports.MAILER_REQUEST_VALUES.valid, { templateName: exports.MAILER_REQUEST_VALUES.invalid.templateName }),
    INVALID_MESSAGE_PARAMS: utils_1.override(exports.MAILER_REQUEST_VALUES.valid, { messageParams: exports.MAILER_REQUEST_VALUES.invalid.messageParams }),
};
exports.MAILER_REQUESTS = {
    valid: exports.MAILER_REQUESTS_VALID,
    invalid: exports.MAILER_REQUESTS_INVALID,
};
