// @comicrelief/data-models/schemas/user-service/models/index.js
"use strict";
/**
 * Re-export all models for ease of use
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDestination = exports.setOrigin = exports.UserMerge = exports.User = exports.Identity = exports.IdentityBase = exports.IdentityContent = exports.IdentityKind = exports.IdentityType = exports.addIdentity = exports.addUser = exports.ErpPayload = exports.ErpPayloadTypes = exports.SignupRequest = exports.PartialSignupRequest = exports.MatchRequest = void 0;
var MatchRequest_1 = require("./requests/MatchRequest");
Object.defineProperty(exports, "MatchRequest", { enumerable: true, get: function () { return MatchRequest_1.MatchRequest; } });
var SignupRequest_1 = require("./requests/SignupRequest");
Object.defineProperty(exports, "PartialSignupRequest", { enumerable: true, get: function () { return SignupRequest_1.PartialSignupRequest; } });
Object.defineProperty(exports, "SignupRequest", { enumerable: true, get: function () { return SignupRequest_1.SignupRequest; } });
var ErpPayload_1 = require("./ErpPayload");
Object.defineProperty(exports, "ErpPayloadTypes", { enumerable: true, get: function () { return ErpPayload_1.ErpPayloadTypes; } });
Object.defineProperty(exports, "ErpPayload", { enumerable: true, get: function () { return ErpPayload_1.ErpPayload; } });
Object.defineProperty(exports, "addUser", { enumerable: true, get: function () { return ErpPayload_1.addUser; } });
Object.defineProperty(exports, "addIdentity", { enumerable: true, get: function () { return ErpPayload_1.addIdentity; } });
var Identity_1 = require("./Identity");
Object.defineProperty(exports, "IdentityType", { enumerable: true, get: function () { return Identity_1.IdentityType; } });
Object.defineProperty(exports, "IdentityKind", { enumerable: true, get: function () { return Identity_1.IdentityKind; } });
Object.defineProperty(exports, "IdentityContent", { enumerable: true, get: function () { return Identity_1.IdentityContent; } });
Object.defineProperty(exports, "IdentityBase", { enumerable: true, get: function () { return Identity_1.IdentityBase; } });
Object.defineProperty(exports, "Identity", { enumerable: true, get: function () { return Identity_1.Identity; } });
var User_1 = require("./User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return User_1.User; } });
var UserMerge_1 = require("./UserMerge");
Object.defineProperty(exports, "UserMerge", { enumerable: true, get: function () { return UserMerge_1.UserMerge; } });
Object.defineProperty(exports, "setOrigin", { enumerable: true, get: function () { return UserMerge_1.setOrigin; } });
Object.defineProperty(exports, "setDestination", { enumerable: true, get: function () { return UserMerge_1.setDestination; } });
