// @comicrelief/data-models/schemas/erp/models/actions/justgiving/DonationPut.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DonationPut = void 0;
var yup = __importStar(require("yup"));
var StrictString_1 = require("../../../../../fields/StrictString");
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
exports.DonationPut = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'DonationPut', function () { return yup.object({
    type: yup.string().oneOf([common_1.ErpPayloadTypes.JUSTGIVING__DONATION__PUT]).required(),
    data: yup.object().shape({
        supporter_uuid: yup.string().uuid().nullable(),
        id: yup.number(),
        amount: yup.number(),
        giftaid: yup.number().integer().nullable(),
        justgivingPage: yup.string().nullable().default(function () { return null; }).defined(),
        isAnonymous: yup.boolean(),
        address1: StrictString_1.StrictString.nullable(),
        address2: StrictString_1.StrictString.nullable(),
        country: StrictString_1.StrictString.nullable(),
        county: StrictString_1.StrictString.nullable(),
        postcode: StrictString_1.StrictString.nullable(),
        paymentFrequency: yup.string().nullable().default(function () { return null; }).defined(),
    }).required(),
}); });
