// @comicrelief/data-models/schemas/erp/models/actions/justgiving/GivingPagePut.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GivingPagePut = void 0;
var yup = __importStar(require("yup"));
var StrictString_1 = require("../../../../../fields/StrictString");
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
/* eslint-disable no-template-curly-in-string */
/**
 * GivingPagePut ERP action model.
 */
exports.GivingPagePut = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'GivingPagePut', function () { return yup.object({
    type: yup.string().oneOf([common_1.ErpPayloadTypes.JUSTGIVING__GIVINGPAGE__PUT]).required(),
    data: yup.object().shape({
        supporter_uuid: validation_1.SupporterUUID.defined(),
        id: yup.string(),
        url: StrictString_1.StrictString.url().defined(),
        title: yup.string(),
        targetAmount: yup.number(),
        totalDonations: yup.number(),
        totalDonationsPlusGiftAid: yup.number(),
        totalSmsDonations: yup.number(),
        offlineAmount: yup.number().required(),
        eventName: yup.string().required(),
        eventId: yup.number().required(),
        charity: yup.number().required(),
        emailConsent: yup.boolean().nullable().default(function () { return null; }).defined(),
        postConsent: yup.boolean().nullable().default(function () { return null; }).defined(),
        addressLine1: yup.string().default(function () { return ''; }).defined(),
        addressLine2: yup.string().default(function () { return ''; }).defined(),
        town: yup.string().default(function () { return ''; }).defined(),
        postcode: yup.string().default(function () { return ''; }).defined(),
        region: yup.string().default(function () { return ''; }).defined(),
        country: yup.string().default(function () { return ''; }).defined(),
        firstName: yup.string().default(function () { return ''; }).defined(),
        lastName: yup.string().default(function () { return ''; }).defined(),
        email: StrictString_1.StrictString.email().defined(),
    }).required(),
}); });
