// @comicrelief/data-models/schemas/prize-platform/mocks/PrizeCategory.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prizeCategoryGenerator = exports.PRIZE_CATEGORIES = exports.PRIZE_CATEGORIES_INVALID = exports.PRIZE_CATEGORIES_VALID = exports.PRIZE_CATEGORY_VALUES = exports.PRIZE_CATEGORY_VALUES_INVALID = exports.PRIZE_CATEGORY_VALUES_VALID = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../utils");
exports.PRIZE_CATEGORY_VALUES_VALID = {
    id: '3fd5502b-3c7b-4b2c-a6f9-0021f5690681',
    name: 'My Descriptive Category',
};
exports.PRIZE_CATEGORY_VALUES_INVALID = {
    id: null,
    name: null,
};
exports.PRIZE_CATEGORY_VALUES = {
    valid: exports.PRIZE_CATEGORY_VALUES_VALID,
    invalid: exports.PRIZE_CATEGORY_VALUES_INVALID,
};
exports.PRIZE_CATEGORIES_VALID = {
    DEFAULT: exports.PRIZE_CATEGORY_VALUES.valid,
};
exports.PRIZE_CATEGORIES_INVALID = {
    INVALID_ID: utils_1.override(exports.PRIZE_CATEGORY_VALUES.valid, { id: exports.PRIZE_CATEGORY_VALUES.invalid.id }),
    INVALID_NAME: utils_1.override(exports.PRIZE_CATEGORY_VALUES.valid, { name: exports.PRIZE_CATEGORY_VALUES.invalid.name }),
    MISSING_UUID: utils_1.omit(exports.PRIZE_CATEGORY_VALUES.valid, ['id']),
    MISSING_NAME: utils_1.omit(exports.PRIZE_CATEGORY_VALUES.valid, ['name']),
};
exports.PRIZE_CATEGORIES = {
    invalid: exports.PRIZE_CATEGORIES_INVALID,
    valid: exports.PRIZE_CATEGORIES_VALID,
};
/**
 * Generates a random prize
 *
 * @param overrides
 */
var prizeCategoryGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ id: uuid_1.v4(), name: utils_1.getString(10) }, overrides));
};
exports.prizeCategoryGenerator = prizeCategoryGenerator;
