// @comicrelief/data-models/schemas/erp/models/doctypes/psl/Payin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payin = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var StrictString_1 = require("../../../../../fields/StrictString");
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
var PayinActivity_1 = require("./PayinActivity");
/**
 * Name of this doctype in ERP.
 */
exports.DOCTYPE = common_1.ErpResources.PAYIN;
/**
 * Model for a `Payin` doctype in ERP.
 */
exports.Payin = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    id: yup.string().uuid().required(),
    transaction_id: yup.string().uuid().required(),
    organisation_uuid: yup.string().uuid().nullable().default(function () { return null; })
        .defined(),
    postal_id: yup.string().required(),
    created: validation_1.TimestampOrNow.required(),
    updated: validation_1.TimestampOrNow.required(),
    fundraising_segment: StrictString_1.StrictString.required(),
    what_you_did: yup.string().default(function () { return ''; }).defined(),
    phone: yup.string().default(function () { return ''; }).defined(),
    mobile: yup.string().default(function () { return ''; }).defined(),
    fundraising_activities: yup.array(PayinActivity_1.ErpPayinActivity.schema().required()).optional(),
}); });
