// @comicrelief/data-models/schemas/erp/models/index.js
"use strict";
/**
 * Schema for `serverless-erp-service`
 *
 * ========================================================
 * GUIDELINES FOR NEW MODELS
 *
 * Actions added to this schema should be
 * added to `./requests/ProducerRequest`.
 * Their mocks should also be added to the
 * relevant ProducerRequest mocks.
 *
 * See the following for details:
 * - src/schemas/erp/models/requests/ProducerRequest.js
 * - src/schemas/erp/mocks/requests/ProducerRequest.mock.js
 * ========================================================
 *
 * In this file, we re-export all models for ease of use
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DonationPut = exports.toFundraiserSignupIdea = exports.FundraisingActivity = exports.toFundraiserSignupTool = exports.FundraisingTool = exports.JustGivingPage = exports.SupporterAddress = exports.SupporterOrganisation = exports.FundraisingSegment = exports.PostalDocType = exports.Postal = exports.Organisation = exports.ErpFundraiserSignup = exports.ErpShopifyOrderLineItem = exports.ErpShopifyOrder = exports.ErpProductVariant = exports.ErpProduct = exports.Payin = exports.RegularGiving = exports.Donation = exports.PostalTicket = exports.PostalTicketStates = exports.Ticket = exports.Prize = exports.ProducerRequest = exports.FundraiserSignupSubmit = exports.UserCreate = exports.IdentityAdd = exports.PayinSubmit = exports.GiftaidUpdate = exports.GiftaidSubmit = exports.TransactionExecuteIvr = exports.TransactionSmsPut = exports.TransactionExecute = exports.SubscriptionUpdate = exports.SubscriptionExecute = exports.ShopifyOrderPut = exports.PayinPut = exports.TicketPut = exports.PostalTicketSetState = exports.PostalTicketPut = exports.PrizePut = void 0;
// -------- ACTIONS --------
var PrizePut_1 = require("./actions/prize/PrizePut");
Object.defineProperty(exports, "PrizePut", { enumerable: true, get: function () { return PrizePut_1.PrizePut; } });
var PostalTicketPut_1 = require("./actions/prize/PostalTicketPut");
Object.defineProperty(exports, "PostalTicketPut", { enumerable: true, get: function () { return PostalTicketPut_1.PostalTicketPut; } });
var PostalTicketSetState_1 = require("./actions/prize/PostalTicketSetState");
Object.defineProperty(exports, "PostalTicketSetState", { enumerable: true, get: function () { return PostalTicketSetState_1.PostalTicketSetState; } });
var TicketPut_1 = require("./actions/prize/TicketPut");
Object.defineProperty(exports, "TicketPut", { enumerable: true, get: function () { return TicketPut_1.TicketPut; } });
var PayinPut_1 = require("./actions/psl/PayinPut");
Object.defineProperty(exports, "PayinPut", { enumerable: true, get: function () { return PayinPut_1.PayinPut; } });
var OrderPut_1 = require("./actions/shopify/OrderPut");
Object.defineProperty(exports, "ShopifyOrderPut", { enumerable: true, get: function () { return OrderPut_1.ShopifyOrderPut; } });
var SubscriptionExecute_1 = require("./actions/SubscriptionExecute");
Object.defineProperty(exports, "SubscriptionExecute", { enumerable: true, get: function () { return SubscriptionExecute_1.SubscriptionExecute; } });
var SubscriptionUpdate_1 = require("./actions/SubscriptionUpdate");
Object.defineProperty(exports, "SubscriptionUpdate", { enumerable: true, get: function () { return SubscriptionUpdate_1.SubscriptionUpdate; } });
var TransactionExecute_1 = require("./actions/TransactionExecute");
Object.defineProperty(exports, "TransactionExecute", { enumerable: true, get: function () { return TransactionExecute_1.TransactionExecute; } });
var TransactionSmsPut_1 = require("./actions/TransactionSmsPut");
Object.defineProperty(exports, "TransactionSmsPut", { enumerable: true, get: function () { return TransactionSmsPut_1.TransactionSmsPut; } });
var TransactionExecuteIvr_1 = require("./actions/TransactionExecuteIvr");
Object.defineProperty(exports, "TransactionExecuteIvr", { enumerable: true, get: function () { return TransactionExecuteIvr_1.TransactionExecuteIvr; } });
var GiftaidSubmit_1 = require("./actions/GiftaidSubmit");
Object.defineProperty(exports, "GiftaidSubmit", { enumerable: true, get: function () { return GiftaidSubmit_1.GiftaidSubmit; } });
var GiftaidUpdate_1 = require("./actions/GiftaidUpdate");
Object.defineProperty(exports, "GiftaidUpdate", { enumerable: true, get: function () { return GiftaidUpdate_1.GiftaidUpdate; } });
var PayinSubmit_1 = require("./actions/PayinSubmit");
Object.defineProperty(exports, "PayinSubmit", { enumerable: true, get: function () { return PayinSubmit_1.PayinSubmit; } });
var IdentityAdd_1 = require("./actions/IdentityAdd");
Object.defineProperty(exports, "IdentityAdd", { enumerable: true, get: function () { return IdentityAdd_1.IdentityAdd; } });
var UserCreate_1 = require("./actions/UserCreate");
Object.defineProperty(exports, "UserCreate", { enumerable: true, get: function () { return UserCreate_1.UserCreate; } });
var FundraiserSignupSubmit_1 = require("./actions/FundraiserSignupSubmit");
Object.defineProperty(exports, "FundraiserSignupSubmit", { enumerable: true, get: function () { return FundraiserSignupSubmit_1.FundraiserSignupSubmit; } });
// -------- REQUESTS --------
var ProducerRequest_1 = require("./requests/ProducerRequest");
Object.defineProperty(exports, "ProducerRequest", { enumerable: true, get: function () { return ProducerRequest_1.ProducerRequest; } });
// -------- DOCTYPES --------
var Prize_1 = require("./doctypes/prize/Prize");
Object.defineProperty(exports, "Prize", { enumerable: true, get: function () { return Prize_1.Prize; } });
var Ticket_1 = require("./doctypes/prize/Ticket");
Object.defineProperty(exports, "Ticket", { enumerable: true, get: function () { return Ticket_1.Ticket; } });
var PostalTicket_1 = require("./doctypes/prize/PostalTicket");
Object.defineProperty(exports, "PostalTicketStates", { enumerable: true, get: function () { return PostalTicket_1.PostalTicketStates; } });
Object.defineProperty(exports, "PostalTicket", { enumerable: true, get: function () { return PostalTicket_1.PostalTicket; } });
var Donation_1 = require("./doctypes/psl/Donation");
Object.defineProperty(exports, "Donation", { enumerable: true, get: function () { return Donation_1.Donation; } });
var RegularGiving_1 = require("./doctypes/psl/RegularGiving");
Object.defineProperty(exports, "RegularGiving", { enumerable: true, get: function () { return RegularGiving_1.RegularGiving; } });
var Payin_1 = require("./doctypes/psl/Payin");
Object.defineProperty(exports, "Payin", { enumerable: true, get: function () { return Payin_1.Payin; } });
var Product_1 = require("./doctypes/shopify/Product");
Object.defineProperty(exports, "ErpProduct", { enumerable: true, get: function () { return Product_1.ErpProduct; } });
var ProductVariant_1 = require("./doctypes/shopify/ProductVariant");
Object.defineProperty(exports, "ErpProductVariant", { enumerable: true, get: function () { return ProductVariant_1.ErpProductVariant; } });
var ShopifyOrder_1 = require("./doctypes/shopify/ShopifyOrder");
Object.defineProperty(exports, "ErpShopifyOrder", { enumerable: true, get: function () { return ShopifyOrder_1.ErpShopifyOrder; } });
var ShopifyOrderLineItem_1 = require("./doctypes/shopify/ShopifyOrderLineItem");
Object.defineProperty(exports, "ErpShopifyOrderLineItem", { enumerable: true, get: function () { return ShopifyOrderLineItem_1.ErpShopifyOrderLineItem; } });
var FundraiserSignup_1 = require("./doctypes/FundraiserSignup");
Object.defineProperty(exports, "ErpFundraiserSignup", { enumerable: true, get: function () { return FundraiserSignup_1.ErpFundraiserSignup; } });
var Organisation_1 = require("./doctypes/Organisation");
Object.defineProperty(exports, "Organisation", { enumerable: true, get: function () { return Organisation_1.Organisation; } });
var Postal_1 = require("./doctypes/Postal");
Object.defineProperty(exports, "Postal", { enumerable: true, get: function () { return Postal_1.Postal; } });
Object.defineProperty(exports, "PostalDocType", { enumerable: true, get: function () { return Postal_1.PostalDocType; } });
var FundraisingSegment_1 = require("./doctypes/FundraisingSegment");
Object.defineProperty(exports, "FundraisingSegment", { enumerable: true, get: function () { return FundraisingSegment_1.FundraisingSegment; } });
var SupporterOrganisation_1 = require("./doctypes/SupporterOrganisation");
Object.defineProperty(exports, "SupporterOrganisation", { enumerable: true, get: function () { return SupporterOrganisation_1.SupporterOrganisation; } });
var SupporterAddress_1 = require("./doctypes/SupporterAddress");
Object.defineProperty(exports, "SupporterAddress", { enumerable: true, get: function () { return SupporterAddress_1.SupporterAddress; } });
var JustGivingPage_1 = require("./doctypes/JustGivingPage");
Object.defineProperty(exports, "JustGivingPage", { enumerable: true, get: function () { return JustGivingPage_1.JustGivingPage; } });
var FundraisingTool_1 = require("./doctypes/FundraisingTool");
Object.defineProperty(exports, "FundraisingTool", { enumerable: true, get: function () { return FundraisingTool_1.FundraisingTool; } });
Object.defineProperty(exports, "toFundraiserSignupTool", { enumerable: true, get: function () { return FundraisingTool_1.toFundraiserSignupTool; } });
var FundraisingActivity_1 = require("./doctypes/FundraisingActivity");
Object.defineProperty(exports, "FundraisingActivity", { enumerable: true, get: function () { return FundraisingActivity_1.FundraisingActivity; } });
Object.defineProperty(exports, "toFundraiserSignupIdea", { enumerable: true, get: function () { return FundraisingActivity_1.toFundraiserSignupIdea; } });
var DonationPut_1 = require("./actions/justgiving/DonationPut");
Object.defineProperty(exports, "DonationPut", { enumerable: true, get: function () { return DonationPut_1.DonationPut; } });
