// @comicrelief/test-utils/UserServiceTester.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserServiceTester = exports.UserServiceEndpoints = void 0;
var axios_1 = __importDefault(require("axios"));
var utils_1 = require("./utils");
var Tester_1 = require("./Tester");
exports.UserServiceEndpoints = {
    MATCH: '/match',
    SIGNUP: '/signup',
};
/**
 * Retrieves and manages
 * entities on `serverless-user-service`
 */
var UserServiceTester = /** @class */ (function (_super) {
    __extends(UserServiceTester, _super);
    /**
     * Constructs an User Service Tester instance.
     *
     * If `apiKey` or `apiUrl` are falsy
     * tries to retrieve them from the environment.
     *
     * Retries defaults are designed to
     * retry the functions for 60 seconds.
     *
     * @param apiKey - defaults to `process.env.TESTS_USER_SERVICE_URL || process.env.USER_SERVICE_URL`
     * @param apiUrl - defaults to `process.env.TESTS_USER_SERVICE_API_KEY || process.env.USER_SERVICE_API_KEY`
     * @param defaultRetries - defaults to 60
     * @param defaultSleep - defaults to 1000 (1 minute)
     */
    function UserServiceTester(apiKey, apiUrl, defaultRetries, defaultSleep) {
        if (apiKey === void 0) { apiKey = ''; }
        if (apiUrl === void 0) { apiUrl = ''; }
        if (defaultRetries === void 0) { defaultRetries = 60; }
        if (defaultSleep === void 0) { defaultSleep = 1000; }
        var _this = _super.call(this) || this;
        _this.apiUrl = apiUrl || process.env.TESTS_USER_SERVICE_URL || process.env.USER_SERVICE_URL || '';
        _this.apiKey = apiKey || process.env.TESTS_USER_SERVICE_API_KEY || process.env.USER_SERVICE_API_KEY || '';
        if (!_this.apiUrl) {
            throw new Error('apiUrl is not defined.');
        }
        if (!_this.apiKey) {
            throw new Error('apiKey is not defined.');
        }
        // Set defaults for 10 minutes
        _this.defaults = {
            retries: defaultRetries,
            sleep: defaultSleep,
        };
        return _this;
    }
    /**
     * Returns a pre-configured
     * axios instance.
     *
     * NB: This just makes tests easier
     */
    UserServiceTester.prototype.getClient = function () {
        return axios_1.default.create({
            baseURL: this.apiUrl,
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': this.apiKey,
            },
        });
    };
    /**
     * Queries UserServiceEndpoints.SIGNUP
     *
     * @param data
     */
    UserServiceTester.prototype.signup = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.getClient().post(exports.UserServiceEndpoints.SIGNUP, data)];
            });
        });
    };
    /**
     * Queries UserServiceEndpoints.MATCH
     *
     * @param data
     * @param retries
     * @param sleep
     */
    UserServiceTester.prototype.match = function (data, retries, sleep) {
        return __awaiter(this, void 0, void 0, function () {
            var attempt, response;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attempt = function () { return _this.getClient().post(exports.UserServiceEndpoints.MATCH, data); };
                        return [4 /*yield*/, utils_1.retry(attempt, retries || this.defaults.retries, sleep || this.defaults.sleep)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * @inheritdoc
     */
    // eslint-disable-next-line class-methods-use-this
    UserServiceTester.prototype.removeReference = function (_reference) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    return UserServiceTester;
}(Tester_1.Tester));
exports.UserServiceTester = UserServiceTester;
