// @comicrelief/data-models/schemas/prize-platform/mocks/requests/tickets/VerifyRequest.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyRequestGenerator = exports.VERIFY_REQUESTS = exports.VERIFY_REQUESTS_INVALID = exports.VERIFY_REQUESTS_VALID = exports.VERIFY_REQUEST_VALUES = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../../../utils");
var Transaction_mock_1 = require("../../Transaction.mock");
var VERIFY_REQUEST_VALID_VALUES = {
    transaction_uuid: Transaction_mock_1.TRANSACTION_VALUES.valid.uuid,
};
var VERIFY_REQUEST_INVALID_VALUES = {
    transaction_uuid: Transaction_mock_1.TRANSACTION_VALUES.invalid.uuid,
};
exports.VERIFY_REQUEST_VALUES = {
    valid: VERIFY_REQUEST_VALID_VALUES,
    invalid: VERIFY_REQUEST_INVALID_VALUES,
};
exports.VERIFY_REQUESTS_VALID = {
    DEFAULT: exports.VERIFY_REQUEST_VALUES.valid,
    ANOTHER: utils_1.override(exports.VERIFY_REQUEST_VALUES.valid, { transaction_uuid: '420b3256-e6c7-48f7-bd0d-5edfca8a132a' }),
};
exports.VERIFY_REQUESTS_INVALID = {
    DEFAULT: exports.VERIFY_REQUEST_VALUES.invalid,
    INVALID_TRANSACTION_UUID: utils_1.override(exports.VERIFY_REQUEST_VALUES.valid, { transaction_uuid: exports.VERIFY_REQUEST_VALUES.invalid.transaction_uuid }),
};
exports.VERIFY_REQUESTS = {
    valid: exports.VERIFY_REQUESTS_VALID,
    invalid: exports.VERIFY_REQUESTS_INVALID,
};
/**
 * Generates a random VerifyRequest
 *
 * @param overrides
 */
var verifyRequestGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ transaction_uuid: uuid_1.v4() }, overrides));
};
exports.verifyRequestGenerator = verifyRequestGenerator;
