// @comicrelief/data-models/schemas/marketing-preferences/transsource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sources = void 0;
/**
 * Known `transSource` values for marketing preferences.
 *
 * This list was obtained by querying the production Contact Store:
 *
 * ```sql
 * SELECT DISTINCT transSource
 * FROM marketing_preferences
 * ORDER BY transSource;
 * ```
 */
exports.sources = {
    /** Unknown -- default catch-all? */
    CR_MARKETINGPREFERENCES: 'CR_MarketingPreferences',
    /** Unknown, lowercase version of `CR_MARKETINGPREFERENCES` */
    CR_MARKETINGPREFERENCES2: 'cr_marketingpreferences',
    /**
     * Obsolete.
     *
     * Used by the first version of the Fundraiser Signup form. Superceded by
     * `FUNDRAISER_SIGNUP_V2` and then `FUNDRAISER_SIGNUP`.
     */
    FUNDRAISING_SIGNUP_FORM: 'fundraising-signup-form',
    /**
     * Obsolete.
     *
     * Used briefly by the second version of the Fundraiser Signup form.
     * Superceded by `FUNDRAISER_SIGNUP`.
     */
    FUNDRAISER_SIGNUP_V2: 'fundraiser-signup-v2',
    /**
     * Fundraiser Signup form (https://www.comicrelief.com/rednoseday/register/)
     *
     * This is the current `transSource` used by FSU since June 2021. Older
     * versions used `FUNDRAISING_SIGNUP_FORM` and `FUNDRAISER_SIGNUP_V2`.
     */
    FUNDRAISER_SIGNUP: 'fundraiser-signup',
    /** Unknown */
    GAMING: 'gaming',
    /**
     * Unknown -- JustGiving?
     *
     * Appears to be related to FROST based on the `transSourceUrl`:
     * https://frost-production.apps.comicrelief.com/registration/create-account
     */
    GIVINGPAGES_REGISTRATION: 'givingpages-registration',
    /** Unknown -- used when making test requests via Postman? */
    POSTMAN: 'postman',
    /** Prize Platform (https://www.comicrelief.com/prizes) */
    PRIZE: 'PRIZE',
    /** Unknown -- used by feature tests, but also appears in production data */
    REACT_MARKETING_PREFERENCES: 'react-marketing-preferences',
    /** Online shop purchase (https://shop.comicrelief.com) */
    SHOP_MARKETINGPREFERENCES: 'CRSHOP_MarketingPreferences',
    /** SR18 donation (https://donation.sportrelief.com) */
    SR18_DONATE: 'SR18_Donate',
    /** RND19 donation (https://donation.rednoseday.com) */
    RND19_DONATE: 'RND19_Donate',
    /** RND19 Gift Aid declaration (https://giftaid.rednoseday.com) */
    RND19_GIFTAID: 'RND19_GiftAid',
    /** SR20 donation (https://donation.sportrelief.com) */
    SR20_DONATE: 'SR20_Donate',
    /** SR20 Gift Aid declaration (https://giftaid.sportrelief.com) */
    SR20_GIFTAID: 'SR20_GiftAid',
    /** Unknown -- other SR20 activity catch-all? */
    SR20_MARKETINGPREFERENCES: 'SR20_MarketingPreferences',
    /** SR20 pay-in (https://donation.sportrelief.com) */
    SR20_PAYIN: 'SR20_Payin',
    /** Big Night In donation (https://donation.bignightin.tv) */
    CV19_DONATE: 'CV19_Donate',
    /** Big Night In Gift Aid declaration (https://giftaid.bignightin.tv) */
    CV19_GIFTAID: 'CV19_GiftAid',
    /** RND21 donation (https://donation.comicrelief.com) */
    RND21_DONATE: 'RND21_Donate',
    /** RND21 Gift Aid declaration (https://giftaid.comicrelief.com) */
    RND21_GIFTAID: 'RND21_GiftAid',
    /** RND21 payin (https://donation.comicrelief.com with `rnd21-payin` cart ID) */
    RND21_PAYIN: 'RND21_Payin',
};
