// @comicrelief/data-models/schemas/user-service/models/requests/SignupRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignupRequest = exports.PartialSignupRequest = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../validation");
var Identity_1 = require("../Identity");
/**
 * SignupRequest model.
 *
 * Specifies the payload to a request
 * to Supporter Events's producer lambda.
 *
 * One and only one between
 * `email` or `mobile`
 * must be provided, and be valid.
 */
exports.PartialSignupRequest = validation_1.defineModel(validation_1.Namespaces.USER_SERVICE, 'PartialSignupRequest', function () {
    var identity = yup.object({
        kind: Identity_1.IdentityKind,
        content: Identity_1.IdentityContent,
    }).required();
    return yup
        .object({
        firstName: validation_1.FirstName.default(function () { return ''; }),
        lastName: validation_1.LastName.default(function () { return ''; }),
        identities: yup.array().of(identity).required(),
    });
});
/**
 * SignupRequest model.
 *
 * Specifies the payload to a request
 * to UserService's signup lambda.
 *
 * One and only one between
 * `email` or `mobile`
 * must be provided, and be valid.
 */
exports.SignupRequest = validation_1.defineModel(validation_1.Namespaces.USER_SERVICE, 'SignupRequest', function () {
    var identity = yup.object({
        hash: yup.string().required(),
        kind: Identity_1.IdentityKind,
        content: Identity_1.IdentityContent,
    }).required();
    return yup
        .object({
        firstName: validation_1.FirstName.default(function () { return ''; }),
        lastName: validation_1.LastName.default(function () { return ''; }),
        identities: yup.array().of(identity).required(),
    });
});
