// @comicrelief/data-models/schemas/erp/models/doctypes/FundraiserSignup.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErpFundraiserSignup = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var enums_1 = require("../../../../enums");
var PhoneNumber_1 = require("../../../../fields/PhoneNumber");
var StrictString_1 = require("../../../../fields/StrictString");
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
var FundraiserSignupIdea_1 = require("./fsu/FundraiserSignupIdea");
var FundraiserSignupTool_1 = require("./fsu/FundraiserSignupTool");
exports.DOCTYPE = common_1.ErpResources.FUNDRAISER_SIGNUP;
exports.ErpFundraiserSignup = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    uuid: StrictString_1.StrictString.uuid().required(),
    supporter_uuid: StrictString_1.StrictString.uuid().required(),
    organisation_uuid: StrictString_1.StrictString.uuid().nullable().defined(),
    postal_id: StrictString_1.StrictString.required(),
    phone: PhoneNumber_1.PhoneNumber.nullable().default(null).defined(),
    campaign_code: StrictString_1.StrictString.required().oneOf(Object.values(enums_1.Campaigns)),
    fundraising_segment: StrictString_1.StrictString.required(),
    target: StrictString_1.StrictString.defined().ensure(),
    justgiving_page_id: yup.number().strict(true).integer().positive()
        .defined()
        .nullable(),
    submitted: validation_1.Timestamp.required(),
    fundraising_tools: yup.array(FundraiserSignupTool_1.ErpFundraiserSignupTool.schema().required()).optional(),
    fundraising_ideas: yup.array(FundraiserSignupIdea_1.ErpFundraiserSignupIdea.schema().required()).optional(),
}); });
