// @comicrelief/data-models/schemas/erp/models/doctypes/FundraisingSegment.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FundraisingSegment = exports.DOCTYPE = void 0;
var lodash_1 = __importDefault(require("lodash"));
var yup = __importStar(require("yup"));
var enums_1 = require("../../../../enums");
var StrictString_1 = require("../../../../fields/StrictString");
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
exports.DOCTYPE = common_1.ErpResources.FUNDRAISING_SEGMENT;
/**
 * The FundraisingSegment doctype is essentially a sub category of
 *  FundraisingGroup.
 *
 * It is a replacement for OrganisationType, which wasn't quite generic
 *  enough, as some values weren't really organisation types. Having said that,
 *  the majority of FundraisingSegment values will be organisation types
 *  (e.g. secondary_school, youth_group etc.)
 *
 * ERP entries (E.g. organisations, fundraiser signups, payin records) should
 *  always be linked to FundraisingSegment records rather than FundraisingGroup
 *  ones. Their FundraisingGroup can be derived from this.
 */
exports.FundraisingSegment = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    label: StrictString_1.StrictString.required(),
    code: StrictString_1.StrictString.test({
        name: 'is-snake-case',
        test: function (value) { return value === lodash_1.default.snakeCase(value); },
        message: '${path} must be in snake case', // eslint-disable-line no-template-curly-in-string
    }).required(),
    fundraising_group: StrictString_1.StrictString.required().oneOf(Object.values(enums_1.FundraisingGroups)),
}); });
