// @comicrelief/data-models/schemas/erp/models/doctypes/FundraisingActivity.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toFundraiserSignupIdea = exports.FundraisingActivity = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var enums_1 = require("../../../../enums");
var StrictString_1 = require("../../../../fields/StrictString");
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
exports.DOCTYPE = common_1.ErpResources.FUNDRAISING_ACTIVITY;
exports.FundraisingActivity = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    code: StrictString_1.StrictString.required().oneOf(Object.values(enums_1.FundraisingActivities)),
    label: StrictString_1.StrictString.required(),
}); });
/**
 * Create a Fundraiser Signup Idea entry from a Fundraising Activity code.
 *
 * Fundraiser Signup Idea is a 'child table' doc type used to store
 *  relationships between Fundraiser Signup entries and Fundraising Activity
 *  entries.
 *
 * See: https://docs.erpnext.com/docs/user/manual/en/customize-erpnext/articles/table-multiselect-field
 *
 * @param code
 * @param fundraiserUuid
 */
var toFundraiserSignupIdea = function (code, fundraiserUuid) { return ({
    doctype: common_1.ErpResources.FUNDRAISER_SIGNUP_IDEA,
    parent: fundraiserUuid,
    parentfield: 'fundraising_ideas',
    parenttype: common_1.ErpResources.FUNDRAISER_SIGNUP,
    code: code,
}); };
exports.toFundraiserSignupIdea = toFundraiserSignupIdea;
