// @comicrelief/data-models/schemas/erp/models/doctypes/SupporterOrganisation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupporterOrganisation = exports.DOCTYPE = void 0;
var uuid_1 = require("uuid");
var yup = __importStar(require("yup"));
var enums_1 = require("../../../../enums");
var StrictString_1 = require("../../../../fields/StrictString");
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
exports.DOCTYPE = common_1.ErpResources.SUPPORTER_ORGANISATION;
/**
 * This model is used to record associations between supporters and organisations.
 *
 * (E.g. the supporter is a teacher and the org is a school.)
 *
 * If the relationship comes to an end, e.g. the supporter leaves the org, the `end` value can be set.
 *
 * source and source_id are intended to record metadata that may be useful, if we have it. source_id isn't
 *  mandatory as we might not have one.
 */
exports.SupporterOrganisation = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    uuid: StrictString_1.StrictString.uuid().defined().nullable(false).default(uuid_1.v4),
    supporter_uuid: StrictString_1.StrictString.uuid().required(),
    organisation_uuid: StrictString_1.StrictString.uuid().required(),
    job_title: StrictString_1.StrictString.default('').ensure(),
    start: validation_1.Timestamp.default(''),
    end: validation_1.Timestamp.default(''),
    // If you are using this doc type, make sure you add your source to the enum.
    source: StrictString_1.StrictString.oneOf(Object.values(enums_1.Sources)).required(),
    source_id: StrictString_1.StrictString.default(''),
}); });
