// @comicrelief/data-models/schemas/erp/mocks/actions/IdentityAdd.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.identityAddGenerator = exports.IDENTITY_ADDS = exports.IDENTITY_ADDS_INVALID = exports.IDENTITY_ADDS_VALID = exports.IDENTITY_ADD_VALUES = exports.IDENTITY_ADD_VALUES_VALID = exports.IDENTITY_ADD_VALUES_INVALID = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../../utils");
var SupporterInfo_mock_1 = require("../../../shared/mocks/SupporterInfo.mock");
var common_1 = require("../../models/common");
exports.IDENTITY_ADD_VALUES_INVALID = {
    type: 'some-type',
    cognito_user_id: null,
    email: [
        'abc',
    ],
    mobile: [
        'efg',
    ],
};
exports.IDENTITY_ADD_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.IDENTITY__ADD,
    cognito_user_id: 'akkwofk!we=',
    email: [
        {
            id: 'cbbdcc1c-798e-440c-8061-0fcdd9c28d35',
            email: SupporterInfo_mock_1.EMAILS.valid,
            hidden: false,
            verified: false,
        },
    ],
    mobile: [
        {
            id: 'd5da8137-059c-40e6-8f9c-6f1c9bb01ec1',
            number: SupporterInfo_mock_1.MOBILES.valid.DEFAULT,
            hidden: false,
            verified: false,
        },
    ],
};
exports.IDENTITY_ADD_VALUES = {
    valid: exports.IDENTITY_ADD_VALUES_VALID,
    invalid: exports.IDENTITY_ADD_VALUES_INVALID,
};
exports.IDENTITY_ADDS_VALID = {
    DEFAULT: exports.IDENTITY_ADD_VALUES.valid,
    EMAIL_ONLY: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { mobile: [] }),
    MOBILE_ONLY: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { email: [] }),
    EMAIL_AND_MOBILE: exports.IDENTITY_ADD_VALUES.valid,
};
exports.IDENTITY_ADDS_INVALID = {
    DEFAULT: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { cognito_user_id: exports.IDENTITY_ADD_VALUES.invalid.cognito_user_id }),
    INVALID_TYPE: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { type: exports.IDENTITY_ADD_VALUES.invalid.type }),
    INVALID_COGNITO_USER_ID: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { cognito_user_id: exports.IDENTITY_ADD_VALUES.invalid.cognito_user_id }),
    INVALID_EMAIL: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { email: exports.IDENTITY_ADD_VALUES.invalid.email }),
    INVALID_MOBILE: utils_1.override(exports.IDENTITY_ADD_VALUES.valid, { mobile: exports.IDENTITY_ADD_VALUES.invalid.mobile }),
    MISSING_TYPE: utils_1.omit(exports.IDENTITY_ADD_VALUES.valid, ['type']),
    MISSING_COGNITO_USER_ID: utils_1.omit(exports.IDENTITY_ADD_VALUES.valid, ['cognito_user_id']),
    MISSING_EMAIL_AND_MOBILE: utils_1.omit(exports.IDENTITY_ADD_VALUES.valid, ['email', 'mobile']),
};
exports.IDENTITY_ADDS = {
    valid: exports.IDENTITY_ADDS_VALID,
    invalid: exports.IDENTITY_ADDS_INVALID,
};
/**
 * Generates a random ProducerRequest
 *
 * @param overrides
 */
var identityAddGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.IDENTITY__ADD, cognito_user_id: uuid_1.v4(), email: [
            {
                id: utils_1.getString(16),
                email: utils_1.getEmail(),
                hidden: false,
                verified: false,
            },
        ], mobile: [
            {
                id: utils_1.getString(16),
                number: utils_1.getMobile(),
                hidden: false,
                verified: false,
            },
        ] }, overrides));
};
exports.identityAddGenerator = identityAddGenerator;
