// @comicrelief/data-models/schemas/erp/mocks/actions/PostalPut.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPostalPut = exports.POSTAL_PUTS = exports.POSTAL_PUTS_INVALID = exports.POSTAL_PUTS_VALID = exports.POSTAL_PUT_VALUES = exports.POSTAL_PUT_VALUES_INVALID = exports.POSTAL_PUT_VALUES_VALID = void 0;
var utils_1 = require("../../../../utils");
var common_1 = require("../../models/common");
var Postal_mock_1 = require("../doctypes/Postal.mock");
exports.POSTAL_PUT_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.POSTAL__PUT,
    data: Postal_mock_1.POSTALS.valid.DEFAULT,
};
exports.POSTAL_PUT_VALUES_INVALID = {
    type: 'some-type',
    data: Postal_mock_1.POSTALS.invalid.DEFAULT,
};
exports.POSTAL_PUT_VALUES = {
    valid: exports.POSTAL_PUT_VALUES_VALID,
    invalid: exports.POSTAL_PUT_VALUES_INVALID,
};
exports.POSTAL_PUTS_VALID = __assign({ DEFAULT: exports.POSTAL_PUT_VALUES.valid }, utils_1.overrideFromMocks(exports.POSTAL_PUT_VALUES_VALID, 'data', Postal_mock_1.POSTALS.valid));
exports.POSTAL_PUTS_INVALID = __assign({ DEFAULT: utils_1.override(exports.POSTAL_PUT_VALUES_VALID, { data: exports.POSTAL_PUT_VALUES_INVALID.data }), INVALID_TYPE: utils_1.override(exports.POSTAL_PUT_VALUES_VALID, { type: exports.POSTAL_PUT_VALUES.invalid.type }), INVALID_DATA: utils_1.override(exports.POSTAL_PUT_VALUES_VALID, { data: null }), MISSING_TYPE: utils_1.omit(exports.POSTAL_PUT_VALUES.valid, ['type']), MISSING_DATA: utils_1.omit(exports.POSTAL_PUT_VALUES.valid, ['data']) }, utils_1.overrideFromMocks(exports.POSTAL_PUT_VALUES_VALID, 'data', Postal_mock_1.POSTALS.invalid));
exports.POSTAL_PUTS = {
    invalid: exports.POSTAL_PUTS_INVALID,
    valid: exports.POSTAL_PUTS_VALID,
};
/**
 * Generates a random PostalPut
 *
 * @param overrides
 */
var createPostalPut = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.POSTAL__PUT, data: Postal_mock_1.createPostal() }, overrides));
};
exports.createPostalPut = createPostalPut;
