// @comicrelief/data-models/schemas/erp/mocks/actions/TransactionExecuteIvr.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transactionExecuteIvrGenerator = exports.TRANSACTION_EXECUTES_IVR = exports.TRANSACTION_EXECUTES_IVR_INVALID = exports.TRANSACTION_EXECUTES_IVR_VALID = exports.TRANSACTION_EXECUTE_IVR_VALUES = exports.TRANSACTION_EXECUTE_IVR_VALUES_VALID = exports.TRANSACTION_EXECUTE_IVR_VALUES_INVALID = void 0;
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
var common_1 = require("../../../payments/models/common");
var common_2 = require("../../models/common");
exports.TRANSACTION_EXECUTE_IVR_VALUES_INVALID = {
    type: 'not-a-type',
    data: {
        id: '',
        supporter_uuid: 'not-a-uuid',
        amount: -5,
        currency: null,
        giftaid: null,
        campaign: 'CR',
        client: 1,
        provider: 'not-a-provider',
        providerReference: 0,
        status: 'not-a-status',
        created: 'not-a-number',
        address1: 0,
        address2: 0,
        address3: 0,
        postcode: 0,
        town: 0,
        country: 0,
    },
};
exports.TRANSACTION_EXECUTE_IVR_VALUES_VALID = {
    type: common_2.ErpPayloadTypes.TRANSACTION__EXECUTE_IVR,
    data: {
        id: '2fadac8d-02c3-4e6b-bbf7-cd4e75bd49d8',
        supporter_uuid: '389d0996-4682-43fe-b90a-f53855dcf7dc',
        amount: 5,
        currency: 'GBP',
        giftaid: 0,
        campaign: 'CR',
        client: 'test',
        provider: common_1.TransactionProvider.UPWARD_COMMS,
        providerReference: 'D00001',
        status: common_1.TransactionStatus.PENDING,
        created: 1603660247,
        address1: null,
        address2: null,
        address3: null,
        postcode: null,
        town: null,
        country: null,
    },
};
var overrideData = function (overridedData) { return ({
    type: common_2.ErpPayloadTypes.TRANSACTION__EXECUTE_IVR,
    data: utils_1.override(exports.TRANSACTION_EXECUTE_IVR_VALUES_VALID.data, overridedData),
}); };
var omitData = function (key) { return ({
    type: common_2.ErpPayloadTypes.TRANSACTION__EXECUTE_IVR,
    data: utils_1.omit(exports.TRANSACTION_EXECUTE_IVR_VALUES_VALID.data, [key]),
}); };
exports.TRANSACTION_EXECUTE_IVR_VALUES = {
    valid: exports.TRANSACTION_EXECUTE_IVR_VALUES_VALID,
    invalid: exports.TRANSACTION_EXECUTE_IVR_VALUES_INVALID,
};
exports.TRANSACTION_EXECUTES_IVR_VALID = {
    DEFAULT: exports.TRANSACTION_EXECUTE_IVR_VALUES.valid,
    UPPERCASE_PROVIDER: overrideData({ provider: common_1.TransactionProvider.DUMMY.toUpperCase() }),
};
exports.TRANSACTION_EXECUTES_IVR_INVALID = {
    DEFAULT: utils_1.override(exports.TRANSACTION_EXECUTE_IVR_VALUES.valid, { data: exports.TRANSACTION_EXECUTE_IVR_VALUES.invalid.data }),
    INVALID_TYPE: utils_1.override(exports.TRANSACTION_EXECUTE_IVR_VALUES.valid, { type: exports.TRANSACTION_EXECUTE_IVR_VALUES.invalid.type }),
    INVALID_ID: overrideData({ id: null }),
    INVALID_SUPPORTER_UUID: overrideData({ supporter_uuid: 'not-a-uuid' }),
    INVALID_AMOUNT: overrideData({ amount: -2 }),
    INVALID_CURRENCY: overrideData({ currency: null }),
    INVALID_CLIENT: overrideData({ client: null }),
    INVALID_PROVIDER: overrideData({ provider: null }),
    INVALID_CREATED: overrideData({ created: null }),
    INVALID_STATUS: overrideData({ status: null }),
    MISSING_ID: omitData('id'),
    MISSING_AMOUNT: omitData('amount'),
    MISSING_CURRENCY: omitData('currency'),
    MISSING_CLIENT: omitData('client'),
    MISSING_PROVIDER: omitData('provider'),
    MISSING_CREATED: omitData('created'),
    MISSING_STATUS: omitData('status'),
    MISSING_SUPPORTER_UUID: omitData('supporter_uuid'),
};
exports.TRANSACTION_EXECUTES_IVR = {
    valid: exports.TRANSACTION_EXECUTES_IVR_VALID,
    invalid: exports.TRANSACTION_EXECUTES_IVR_INVALID,
};
/**
 * Generates a random TransactionExecuteIvr
 *
 * @param overrides
 */
var transactionExecuteIvrGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    var date = Math.floor(utils_1.getUnixTimestamp() / 1000);
    return {
        type: common_2.ErpPayloadTypes.TRANSACTION__EXECUTE_IVR,
        data: __assign({ id: utils_1.getUuid(), supporter_uuid: utils_1.getUuid(), amount: Math.floor(Math.random() * 10) + 1, currency: utils_1.getRandomElement(['GBP', 'USD']), giftaid: 0, campaign: utils_1.getRandomElement([enums_1.Campaigns.RND21, enums_1.Campaigns.SR20]), client: utils_1.getRandomElement(['test', 'donation']), provider: utils_1.getRandomElement(common_1.TransactionProvider), providerReference: 'D12345', status: utils_1.getRandomElement(common_1.TransactionStatus), created: date, address1: utils_1.getRandomElement([utils_1.getString(10), null]), address2: utils_1.getRandomElement([utils_1.getString(10), null]), address3: utils_1.getRandomElement([utils_1.getString(10), null]), postcode: utils_1.getRandomElement([utils_1.getString(10), null]), town: utils_1.getRandomElement([utils_1.getString(10), null]), country: utils_1.getRandomElement([utils_1.getString(10), null]) }, overrides),
    };
};
exports.transactionExecuteIvrGenerator = transactionExecuteIvrGenerator;
