// @comicrelief/data-models/schemas/erp/mocks/actions/UserCreate.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.userCreateGenerator = exports.USER_CREATES = exports.USER_CREATES_INVALID = exports.USER_CREATES_VALID = exports.USER_CREATE_VALUES = exports.USER_CREATE_VALUES_VALID = exports.USER_CREATE_VALUES_INVALID = void 0;
var utils_1 = require("../../../../utils");
var SupporterInfo_mock_1 = require("../../../shared/mocks/SupporterInfo.mock");
var common_1 = require("../../models/common");
var IdentityAdd_mock_1 = require("./IdentityAdd.mock");
exports.USER_CREATE_VALUES_INVALID = {
    type: 'some-type',
    cognito_user_id: null,
    first_name: SupporterInfo_mock_1.SUPPORTER_INFO_VALUES.invalid.firstName,
    last_name: SupporterInfo_mock_1.SUPPORTER_INFO_VALUES.invalid.lastName,
    email: [
        'abc',
    ],
    mobile: [
        'efg',
    ],
};
exports.USER_CREATE_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.USER__CREATE,
    cognito_user_id: 'akkwofk!we=',
    first_name: SupporterInfo_mock_1.SUPPORTER_INFO_VALUES.valid.firstName,
    last_name: SupporterInfo_mock_1.SUPPORTER_INFO_VALUES.valid.lastName,
    email: [
        {
            id: 'cbbdcc1c-798e-440c-8061-0fcdd9c28d35',
            email: SupporterInfo_mock_1.EMAILS.valid,
            hidden: false,
            verified: false,
        },
    ],
    mobile: [
        {
            id: 'd5da8137-059c-40e6-8f9c-6f1c9bb01ec1',
            number: SupporterInfo_mock_1.MOBILES.valid.DEFAULT,
            hidden: false,
            verified: false,
        },
    ],
};
exports.USER_CREATE_VALUES = {
    valid: exports.USER_CREATE_VALUES_VALID,
    invalid: exports.USER_CREATE_VALUES_INVALID,
};
exports.USER_CREATES_VALID = {
    DEFAULT: exports.USER_CREATE_VALUES.valid,
    EMAIL_ONLY: utils_1.override(exports.USER_CREATE_VALUES.valid, { mobile: [] }),
    MOBILE_ONLY: utils_1.override(exports.USER_CREATE_VALUES.valid, { email: [] }),
    EMAIL_AND_MOBILE: exports.USER_CREATE_VALUES.valid,
    MISSING_FIRST_NAME: utils_1.omit(exports.USER_CREATE_VALUES.valid, ['first_name']),
    MISSING_LAST_NAME: utils_1.omit(exports.USER_CREATE_VALUES.valid, ['last_name']),
};
exports.USER_CREATES_INVALID = {
    DEFAULT: utils_1.override(exports.USER_CREATE_VALUES.valid, { cognito_user_id: exports.USER_CREATE_VALUES.invalid.cognito_user_id }),
    INVALID_TYPE: utils_1.override(exports.USER_CREATE_VALUES.valid, { type: exports.USER_CREATE_VALUES.invalid.type }),
    INVALID_COGNITO_USER_ID: utils_1.override(exports.USER_CREATE_VALUES.valid, { cognito_user_id: exports.USER_CREATE_VALUES.invalid.cognito_user_id }),
    INVALID_FIRST_NAME: utils_1.override(exports.USER_CREATE_VALUES.valid, { first_name: exports.USER_CREATE_VALUES.invalid.first_name }),
    INVALID_LAST_NAME: utils_1.override(exports.USER_CREATE_VALUES.valid, { last_name: exports.USER_CREATE_VALUES.invalid.last_name }),
    INVALID_EMAIL: utils_1.override(exports.USER_CREATE_VALUES.valid, { email: exports.USER_CREATE_VALUES.invalid.email }),
    INVALID_MOBILE: utils_1.override(exports.USER_CREATE_VALUES.valid, { mobile: exports.USER_CREATE_VALUES.invalid.mobile }),
    MISSING_TYPE: utils_1.omit(exports.USER_CREATE_VALUES.valid, ['type']),
    MISSING_COGNITO_USER_ID: utils_1.omit(exports.USER_CREATE_VALUES.valid, ['cognito_user_id']),
    MISSING_EMAIL_AND_MOBILE: utils_1.omit(exports.USER_CREATE_VALUES.valid, ['email', 'mobile']),
};
exports.USER_CREATES = {
    valid: exports.USER_CREATES_VALID,
    invalid: exports.USER_CREATES_INVALID,
};
/**
 * Generates a random UserCreate
 *
 * @param overrides
 */
var userCreateGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign(__assign(__assign({}, IdentityAdd_mock_1.identityAddGenerator()), { type: common_1.ErpPayloadTypes.USER__CREATE, first_name: utils_1.getString(16), last_name: utils_1.getString(16) }), overrides));
};
exports.userCreateGenerator = userCreateGenerator;
