// @comicrelief/data-models/schemas/erp/mocks/actions/shopify/OrderPut.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.orderPutGenerator = exports.ORDER_PUTS = exports.ORDER_PUTS_INVALID = exports.ORDER_PUTS_VALID = exports.ORDER_PUT_VALUES = exports.ORDER_PUT_VALUES_INVALID = exports.ORDER_PUT_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
var common_1 = require("../../../models/common");
var ShopifyOrder_mock_1 = require("../../doctypes/shopify/ShopifyOrder.mock");
exports.ORDER_PUT_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.SHOPIFY__ORDER__PUT,
    data: ShopifyOrder_mock_1.SHOPIFY_ORDERS.valid.DEFAULT,
};
exports.ORDER_PUT_VALUES_INVALID = {
    type: 'some-type',
    data: ShopifyOrder_mock_1.SHOPIFY_ORDERS.invalid.DEFAULT,
};
exports.ORDER_PUT_VALUES = {
    valid: exports.ORDER_PUT_VALUES_VALID,
    invalid: exports.ORDER_PUT_VALUES_INVALID,
};
exports.ORDER_PUTS_VALID = __assign({ DEFAULT: exports.ORDER_PUT_VALUES.valid }, utils_1.overrideFromMocks(exports.ORDER_PUT_VALUES_VALID, 'data', ShopifyOrder_mock_1.SHOPIFY_ORDERS.valid));
exports.ORDER_PUTS_INVALID = __assign({ DEFAULT: utils_1.override(exports.ORDER_PUT_VALUES_VALID, { data: exports.ORDER_PUT_VALUES_INVALID.data }), INVALID_TYPE: utils_1.override(exports.ORDER_PUT_VALUES_VALID, { type: exports.ORDER_PUT_VALUES.invalid.type }), INVALID_DATA: utils_1.override(exports.ORDER_PUT_VALUES_VALID, { data: null }), MISSING_TYPE: utils_1.omit(exports.ORDER_PUT_VALUES.valid, ['type']), MISSING_DATA: utils_1.omit(exports.ORDER_PUT_VALUES.valid, ['data']) }, utils_1.overrideFromMocks(exports.ORDER_PUT_VALUES_VALID, 'data', ShopifyOrder_mock_1.SHOPIFY_ORDERS.invalid));
exports.ORDER_PUTS = {
    invalid: exports.ORDER_PUTS_INVALID,
    valid: exports.ORDER_PUTS_VALID,
};
/**
 * Generates a random OrderPut action.
 *
 * @param overrides
 */
var orderPutGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.SHOPIFY__ORDER__PUT, data: ShopifyOrder_mock_1.shopifyOrderGenerator() }, overrides));
};
exports.orderPutGenerator = orderPutGenerator;
