// @comicrelief/data-models/schemas/erp/models/doctypes/shopify/ProductVariant.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErpProductVariant = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
/**
 * Name of this doctype in ERP.
 */
exports.DOCTYPE = common_1.ErpResources.PRODUCT_VARIANT;
/**
 * Model for a `Product` document in ERP.
 */
exports.ErpProductVariant = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    shopify_id: yup.string().required(),
    sku: validation_1.NonEmptyString.nullable().defined(),
    title: yup.string().required(),
    price: validation_1.ErpCurrencyField.min(0).required(),
    currency: yup.string().required(),
    available: validation_1.NumericBoolField.required(),
    option1: validation_1.NonEmptyString.required(),
    option2: validation_1.NonEmptyString.nullable().defined(),
    option3: validation_1.NonEmptyString.nullable().defined(),
}); });
