// @comicrelief/data-models/schemas/erp/models/doctypes/shopify/Product.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DOCTYPE_METADATA = exports.ErpProduct = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
var ProductVariant_1 = require("./ProductVariant");
/**
 * Name of this doctype in ERP.
 */
exports.DOCTYPE = common_1.ErpResources.PRODUCT;
/**
 * Model for a `Product` document in ERP.
 */
exports.ErpProduct = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    shopify_id: yup.string().required(),
    title: yup.string().required(),
    variants: yup.array(ProductVariant_1.ErpProductVariant.schema().required()).min(1).required(),
    available: validation_1.NumericBoolField.required(),
    option1: validation_1.NonEmptyString.nullable().defined(),
    option2: validation_1.NonEmptyString.nullable().defined(),
    option3: validation_1.NonEmptyString.nullable().defined(),
}); });
exports.DOCTYPE_METADATA = {
    doctype: {
        name: exports.DOCTYPE,
        autoname: 'field:shopify_id',
        title_field: 'title',
    },
    fields: {
        variants: {
            options: 'Product Variant',
        },
        available: {
            default: '0',
        },
    },
};
