// @comicrelief/data-models/schemas/erp/mocks/requests/QueryRequest.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryRequestGenerator = exports.queryRequestFilterGenerator = exports.QUERY_REQUESTS = exports.QUERY_REQUESTS_INVALID = exports.QUERY_REQUESTS_VALID = exports.QUERY_REQUEST_VALUES = exports.QUERY_REQUEST_VALUES_INVALID = exports.QUERY_REQUEST_VALUES_VALID = exports.QUERY_REQUEST_INVALID_FILTER = exports.QUERY_REQUEST_VALID_FILTER = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../../utils");
var common_1 = require("../../models/common");
var QueryRequest_1 = require("../../models/requests/QueryRequest");
exports.QUERY_REQUEST_VALID_FILTER = {
    field: 'first_name',
    operator: QueryRequest_1.QueryFilterOperators.EQUALS,
    value: 'Dante Alighieri',
};
exports.QUERY_REQUEST_INVALID_FILTER = {
    field: null,
    operator: null,
    value: null,
};
exports.QUERY_REQUEST_VALUES_VALID = {
    resource: common_1.ErpResources.SUPPORTER,
    name: '78cb8a41-9280-427e-b9aa-5873c5a359db',
    fields: ['*'],
    filters: [],
    offset: 0,
    limit: 20,
};
exports.QUERY_REQUEST_VALUES_INVALID = {
    resource: 'not-a-resource',
    name: null,
    fields: [null],
    filters: [1],
    offset: 'abc',
    limit: 'abc',
};
exports.QUERY_REQUEST_VALUES = {
    valid: exports.QUERY_REQUEST_VALUES_VALID,
    invalid: exports.QUERY_REQUEST_VALUES_INVALID,
};
exports.QUERY_REQUESTS_VALID = {
    DEFAULT: exports.QUERY_REQUEST_VALUES.valid,
    LIMIT_ZERO: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { limit: 0 }),
    MISSING_NAME: utils_1.omit(exports.QUERY_REQUEST_VALUES.valid, ['name']),
    MISSING_FIELDS: utils_1.omit(exports.QUERY_REQUEST_VALUES.valid, ['fields']),
    MISSING_FILTERS: utils_1.omit(exports.QUERY_REQUEST_VALUES.valid, ['filters']),
    MISSING_OFFSET: utils_1.omit(exports.QUERY_REQUEST_VALUES.valid, ['offset']),
    MISSING_LIMIT: utils_1.omit(exports.QUERY_REQUEST_VALUES.valid, ['limit']),
    WITH_NAME: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { name: '57360b88-c996-4a5d-a07c-8c01cdd425db' }),
    WITH_RESOURCE_SPACE: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { resource: common_1.ErpResources.PRIZE_TICKET }),
    WITH_NAME_SPACE: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { name: 'Id With Space' }),
    WITH_NAME_HASH: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { name: 'Id With # Sign' }),
    WITH_NAME_QUESTION: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { name: 'Id With ? Sign' }),
    WITH_NAME_SLASH: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { name: 'Id With / Sign' }),
    WITH_SUPPORTER_FIELDS: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { fields: ['first_name', 'last_name'] }),
    WITH_SUPPORTER_FILTERS: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { filters: [exports.QUERY_REQUEST_VALID_FILTER] }),
    WITH_PRIZE_TICKET_FILTERS: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { resource: common_1.ErpResources.PRIZE_TICKET, filters: [exports.QUERY_REQUEST_VALID_FILTER] }),
    WITH_IMPLICIT_FILTER_OPERATOR: utils_1.override(exports.QUERY_REQUEST_VALUES.valid, { filters: [{ field: 'a', value: 'b' }] }),
};
exports.QUERY_REQUESTS_INVALID = {
    DEFAULT: exports.QUERY_REQUEST_VALUES.invalid,
    MISSING_RESOURCE: utils_1.omit(exports.QUERY_REQUEST_VALUES_VALID, ['resource']),
    INVALID_RESOURCE: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { resource: exports.QUERY_REQUEST_VALUES_INVALID.resource }),
    INVALID_NAME: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { name: exports.QUERY_REQUEST_VALUES_INVALID.name }),
    INVALID_FIELDS: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { fields: exports.QUERY_REQUEST_VALUES_INVALID.fields }),
    INVALID_FIELDS_CONTAINS_NULL: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { fields: ['field', null, 'another'] }),
    INVALID_FILTERS: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { filters: exports.QUERY_REQUEST_VALUES_INVALID.filters }),
    INVALID_FILTERS_FORMAT: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { filters: { field: 'a' } }),
    INVALID_FILTERS_VALUES: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { filters: exports.QUERY_REQUEST_INVALID_FILTER }),
    INVALID_OFFSET: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { offset: exports.QUERY_REQUEST_VALUES.invalid.offset }),
    INVALID_LIMIT: utils_1.override(exports.QUERY_REQUEST_VALUES_VALID, { limit: exports.QUERY_REQUEST_VALUES.invalid.limit }),
};
exports.QUERY_REQUESTS = {
    valid: exports.QUERY_REQUESTS_VALID,
    invalid: exports.QUERY_REQUESTS_INVALID,
};
/**
 * Generates a random QueryFilterProps
 *
 * @param overrides
 */
var queryRequestFilterGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    var operator = utils_1.getRandomElement(Object.values(QueryRequest_1.QueryFilterOperators));
    return __assign({ field: utils_1.getString(), operator: operator, value: utils_1.getString() }, overrides);
};
exports.queryRequestFilterGenerator = queryRequestFilterGenerator;
/**
 * Generates a random QueryRequestProps
 *
 * @param overrides
 */
var queryRequestGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    var resource = utils_1.getRandomElement(Object.values(common_1.ErpResources));
    return __assign({ resource: resource, name: uuid_1.v4(), fields: ['*'], filters: [], offset: null, limit: null }, overrides);
};
exports.queryRequestGenerator = queryRequestGenerator;
