// @comicrelief/data-models/schemas/erp/models/actions/GiftaidSubmit.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GiftaidSubmit = void 0;
var yup = __importStar(require("yup"));
var Postcode_1 = require("../../../../fields/Postcode");
var StrictString_1 = require("../../../../fields/StrictString");
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
var GiftAidDeclaration_1 = require("../doctypes/giftaid/GiftAidDeclaration");
/**
 * GiftaidSubmit Action.
 */
exports.GiftaidSubmit = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'GiftaidSubmit', function () {
    var validator = yup.object({
        type: yup.string().oneOf([common_1.ErpPayloadTypes.GIFTAID__SUBMIT]).required(),
        data: yup.object({
            declaration_transaction_id: yup.string().uuid().required(),
            supporter_uuid: yup.string().uuid().nullable().defined(),
            declaration_start_date: validation_1.Timestamp.required(),
            declaration_end_date: validation_1.Timestamp.nullable().default(function () { return null; }).defined(),
            submission_date: validation_1.Timestamp.required(),
            first_name: yup.string().required(),
            last_name: yup.string().required(),
            source: yup.string().oneOf(Object.values(GiftAidDeclaration_1.GiftAidSource)).required(),
            can_claim_giftaid: yup.number().integer().oneOf([0, 1])
                .required(),
            address_line_1: StrictString_1.StrictString.trim().required().max(100),
            address_line_2: yup.string().nullable(),
            address_line_3: yup.string().nullable(),
            town: yup.string().nullable(),
            postcode: Postcode_1.Postcode.required(),
        }).required(),
    });
    return validator;
});
