// @comicrelief/data-models/schemas/erp/models/actions/PayinSubmit.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayinSubmit = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
/**
 * PayinSubmit Action.
 */
exports.PayinSubmit = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'PayinSubmit', function () { return yup.object({
    type: yup.string().oneOf([common_1.ErpPayloadTypes.PAYIN__SUBMIT]).required(),
    data: yup.object({
        id: yup.string().uuid().required(),
        transactionId: yup.string().uuid().required(),
        audience: yup.string().required(),
        postcode: yup.string().required(),
        town: yup.string().required(),
        country: yup.string().required(),
        address1: yup.string().required(),
        address2: yup.string().nullable(),
        address3: yup.string().nullable(),
    }).required(),
}); });
