// @comicrelief/data-models/schemas/erp/models/actions/SubscriptionUpdate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionUpdate = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
/**
 * SubscriptionUpdate Action.
 */
exports.SubscriptionUpdate = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'SubscriptionUpdate', function () {
    var validator = yup.object({
        type: yup.string().oneOf([common_1.ErpPayloadTypes.SUBSCRIPTION__UPDATE]).required(),
        data: yup.object({
            id: yup.string().uuid().required(),
            supporter_uuid: yup.string().uuid().nullable().defined(),
            amount: yup.number().positive().required(),
            billingDayOfMonth: yup.number().positive().required(),
            currency: yup.string().required(),
            client: yup.string().required(),
            provider: yup.string().required(),
            created: yup.number().integer().positive().required(),
            updated: yup.number().integer().positive().required(),
            cancelled: validation_1.Timestamp.nullable().default(function () { return null; }),
            status: yup.string().required(),
            postcode: yup.string().required(),
            town: yup.string().required(),
            country: yup.string().required(),
            address1: yup.string().required(),
            address2: yup.string().nullable(),
            address3: yup.string().nullable(),
            campaign: yup.string().nullable(),
            transSource: yup.string().nullable(),
            transSourceUrl: yup.string().nullable(),
            transType: yup.string().nullable(),
            cartId: yup.string().nullable(),
            providerReference: yup.string().nullable(),
            affiliate: yup.string().nullable(),
            giftaid: yup.number().integer().nullable(),
        }).required(),
    });
    return validator;
});
