// @comicrelief/data-models/schemas/erp/models/actions/UserCreate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserCreate = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../validation");
var common_1 = require("../common");
/**
 * UserCreateAction.
 */
exports.UserCreate = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, 'UserCreate', function () {
    var test = {
        name: 'email-or-mobile',
        message: 'Please provide either \'email\' or \'mobile\'',
        func: function (value) { var _a, _b; return Boolean(value && (((_a = value.email) === null || _a === void 0 ? void 0 : _a.length) || ((_b = value.mobile) === null || _b === void 0 ? void 0 : _b.length))); },
    };
    var validator = yup
        .object({
        type: yup.string().oneOf([common_1.ErpPayloadTypes.USER__CREATE]).required(),
        cognito_user_id: yup.string().required(),
        first_name: validation_1.FirstName.default(function () { return ''; }),
        last_name: validation_1.LastName.default(function () { return ''; }),
        email: yup.array().of(common_1.EmailList),
        mobile: yup.array().of(common_1.MobileList),
    })
        .test(test.name, test.message, test.func);
    return validator;
});
