// @comicrelief/data-models/schemas/erp/mocks/doctypes/shopify/ShopifyOrder.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.shopifyOrderGenerator = exports.SHOPIFY_ORDERS = exports.SHOPIFY_ORDERS_INVALID = exports.SHOPIFY_ORDERS_VALID = exports.SHOPIFY_ORDER_VALUES = exports.SHOPIFY_ORDER_VALUES_INVALID = exports.SHOPIFY_ORDER_VALUES_VALID = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../../../utils");
var validation_1 = require("../../../../../validation");
var ShopifyOrder_1 = require("../../../models/doctypes/shopify/ShopifyOrder");
var ShopifyOrderLineItem_mock_1 = require("./ShopifyOrderLineItem.mock");
exports.SHOPIFY_ORDER_VALUES_VALID = {
    shopify_id: '8579183571029',
    shopify_processed_at: '2020-11-01T12:34:56+00:00',
    shopify_updated_at: '2020-11-02T12:34:56+00:00',
    shopify_cancelled_at: '2020-11-03T12:34:56+00:00',
    shopify_customer_id: '4155130642586',
    supporter_uuid: 'a03add53-1ad6-403b-89b0-7690a7c6ff3d',
    order_number: '#54321',
    cancel_reason: ShopifyOrder_1.CancelledReason.CUSTOMER,
    payment_status: ShopifyOrder_1.PaymentStatus.PAID,
    fulfilment_status: ShopifyOrder_1.FulfilmentStatus.UNFULFILLED,
    postcode: 'SE1 7TP',
    total_line_items_price: 60.00,
    total_discounts: 0.00,
    subtotal_price: 60.00,
    subtotal_includes_vat: validation_1.NumericBool.YES,
    total_outstanding: 0.00,
    total_donation_received: 0.00,
    total_postage: 7.50,
    total_price: 67.50,
    total_vat: 0.00,
    total_refunded: 0.00,
    currency: validation_1.ErpCurrencies.GBP,
    line_items: [ShopifyOrderLineItem_mock_1.LINE_ITEMS.valid.DEFAULT],
};
exports.SHOPIFY_ORDER_VALUES_INVALID = {
    shopify_id: null,
    shopify_processed_at: null,
    shopify_updated_at: null,
    shopify_cancelled_at: 'invalid',
    shopify_customer_id: null,
    supporter_uuid: 'not a uuid',
    order_number: null,
    cancel_reason: 'invalid',
    payment_status: 'blah',
    fulfilment_status: 'blah',
    postcode: null,
    total_line_items_price: -1,
    total_discounts: -1,
    subtotal_price: -1,
    subtotal_includes_vat: 'maybe',
    total_outstanding: -1,
    total_donation_received: -1,
    total_postage: -1,
    total_price: -1,
    total_vat: -1,
    total_refunded: -1,
    currency: '',
    line_items: [],
};
exports.SHOPIFY_ORDER_VALUES = {
    valid: exports.SHOPIFY_ORDER_VALUES_VALID,
    invalid: exports.SHOPIFY_ORDER_VALUES_INVALID,
};
exports.SHOPIFY_ORDERS_VALID = {
    DEFAULT: exports.SHOPIFY_ORDER_VALUES.valid,
    PAYMENT_STATUS_PENDING: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.PENDING }),
    PAYMENT_STATUS_AUTHORIZED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.AUTHORIZED }),
    PAYMENT_STATUS_PARTIALLY_PAID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.PARTIALLY_PAID }),
    PAYMENT_STATUS_PAID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.PAID }),
    PAYMENT_STATUS_PARTIALLY_REFUNDED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.PARTIALLY_REFUNDED }),
    PAYMENT_STATUS_REFUNDED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.REFUNDED }),
    PAYMENT_STATUS_VOIDED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: ShopifyOrder_1.PaymentStatus.VOIDED }),
    FULFILMENT_STATUS_FULFILLED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { fulfilment_status: ShopifyOrder_1.FulfilmentStatus.FULFILLED }),
    FULFILMENT_STATUS_UNFULFILLED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { fulfilment_status: ShopifyOrder_1.FulfilmentStatus.UNFULFILLED }),
    FULFILMENT_STATUS_PARTIAL: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { fulfilment_status: ShopifyOrder_1.FulfilmentStatus.PARTIAL }),
    FULFILMENT_STATUS_RESTOCKED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { fulfilment_status: ShopifyOrder_1.FulfilmentStatus.RESTOCKED }),
    CANCELLED_AT_NULL: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_cancelled_at: null }),
    CANCEL_REASON_NULL: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: null }),
    CANCEL_REASON_CUSTOMER: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: ShopifyOrder_1.CancelledReason.CUSTOMER }),
    CANCEL_REASON_FRAUD: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: ShopifyOrder_1.CancelledReason.FRAUD }),
    CANCEL_REASON_INVENTORY: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: ShopifyOrder_1.CancelledReason.INVENTORY }),
    CANCEL_REASON_DECLINED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: ShopifyOrder_1.CancelledReason.DECLINED }),
    CANCEL_REASON_OTHER: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: ShopifyOrder_1.CancelledReason.OTHER }),
    TWO_LINE_ITEMS: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { line_items: [ShopifyOrderLineItem_mock_1.LINE_ITEMS.valid.DEFAULT, ShopifyOrderLineItem_mock_1.LINE_ITEMS.valid.ALTERNATIVE] }),
    FREE_ORDER: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, {
        total_line_items_price: 0,
        subtotal_price: 0,
        total_postage: 0,
        total_price: 0,
        line_items: [utils_1.override(ShopifyOrderLineItem_mock_1.LINE_ITEMS.valid.DEFAULT, { price: 0 })],
    }),
};
exports.SHOPIFY_ORDERS_INVALID = {
    DEFAULT: exports.SHOPIFY_ORDER_VALUES.invalid,
    INVALID_SHOPIFY_ID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_id: exports.SHOPIFY_ORDER_VALUES.invalid.shopify_id }),
    INVALID_SHOPIFY_PROCESSED_AT: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_processed_at: exports.SHOPIFY_ORDER_VALUES.invalid.shopify_processed_at }),
    INVALID_SHOPIFY_UPDATED_AT: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_updated_at: exports.SHOPIFY_ORDER_VALUES.invalid.shopify_updated_at }),
    INVALID_SHOPIFY_CANCELLED_AT: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_cancelled_at: exports.SHOPIFY_ORDER_VALUES.invalid.shopify_cancelled_at }),
    INVALID_SHOPIFY_CUSTOMER_ID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { shopify_customer_id: exports.SHOPIFY_ORDER_VALUES.invalid.shopify_customer_id }),
    INVALID_SUPPORTER_UUID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { supporter_uuid: exports.SHOPIFY_ORDER_VALUES.invalid.supporter_uuid }),
    INVALID_ORDER_NUMBER: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { order_number: exports.SHOPIFY_ORDER_VALUES.invalid.order_number }),
    INVALID_CANCEL_REASON: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { cancel_reason: exports.SHOPIFY_ORDER_VALUES.invalid.cancel_reason }),
    INVALID_PAYMENT_STATUS: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { payment_status: exports.SHOPIFY_ORDER_VALUES.invalid.payment_status }),
    INVALID_FULFILMENT_STATUS: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { fulfilment_status: exports.SHOPIFY_ORDER_VALUES.invalid.fulfilment_status }),
    INVALID_POSTCODE: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { postcode: exports.SHOPIFY_ORDER_VALUES.invalid.postcode }),
    INVALID_TOTAL_LINE_ITEMS_PRICE: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_line_items_price: exports.SHOPIFY_ORDER_VALUES.invalid.total_line_items_price }),
    INVALID_TOTAL_DISCOUNTS: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_discounts: exports.SHOPIFY_ORDER_VALUES.invalid.total_discounts }),
    INVALID_SUBTOTAL_PRICE: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { subtotal_price: exports.SHOPIFY_ORDER_VALUES.invalid.subtotal_price }),
    INVALID_SUBTOTAL_INCLUDES_VAT: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { subtotal_includes_vat: exports.SHOPIFY_ORDER_VALUES.invalid.subtotal_includes_vat }),
    INVALID_TOTAL_OUTSTANDING: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_outstanding: exports.SHOPIFY_ORDER_VALUES.invalid.total_outstanding }),
    INVALID_TOTAL_DONATION_RECEIVED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_donation_received: exports.SHOPIFY_ORDER_VALUES.invalid.total_donation_received }),
    INVALID_TOTAL_POSTAGE: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_postage: exports.SHOPIFY_ORDER_VALUES.invalid.total_postage }),
    INVALID_TOTAL_PRICE: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_price: exports.SHOPIFY_ORDER_VALUES.invalid.total_price }),
    INVALID_TOTAL_VAT: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_vat: exports.SHOPIFY_ORDER_VALUES.invalid.total_vat }),
    INVALID_TOTAL_REFUNDED: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { total_refunded: exports.SHOPIFY_ORDER_VALUES.invalid.total_refunded }),
    INVALID_CURRENCY: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { currency: exports.SHOPIFY_ORDER_VALUES.invalid.currency }),
    INVALID_LINE_ITEMS: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { line_items: exports.SHOPIFY_ORDER_VALUES.invalid.line_items }),
    MISSING_SHOPIFY_ID: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['shopify_id']),
    MISSING_SHOPIFY_PROCESSED_AT: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['shopify_processed_at']),
    MISSING_SHOPIFY_UPDATED_AT: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['shopify_updated_at']),
    MISSING_SHOPIFY_CANCELLED_AT: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['shopify_cancelled_at']),
    MISSING_SHOPIFY_CUSTOMER_ID: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['shopify_customer_id']),
    MISSING_SUPPORTER_UUID: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['supporter_uuid']),
    MISSING_ORDER_NUMBER: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['order_number']),
    MISSING_CANCEL_REASON: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['cancel_reason']),
    MISSING_PAYMENT_STATUS: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['payment_status']),
    MISSING_FULFILMENT_STATUS: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['fulfilment_status']),
    MISSING_POSTCODE: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['postcode']),
    MISSING_TOTAL_LINE_ITEMS_PRICE: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_line_items_price']),
    MISSING_TOTAL_DISCOUNTS: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_discounts']),
    MISSING_SUBTOTAL_PRICE: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['subtotal_price']),
    MISSING_SUBTOTAL_INCLUDES_VAT: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['subtotal_includes_vat']),
    MISSING_TOTAL_OUTSTANDING: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_outstanding']),
    MISSING_TOTAL_DONATION_RECEIVED: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_donation_received']),
    MISSING_TOTAL_POSTAGE: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_postage']),
    MISSING_TOTAL_PRICE: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_price']),
    MISSING_TOTAL_VAT: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_vat']),
    MISSING_TOTAL_REFUNDED: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['total_refunded']),
    MISSING_CURRENCY: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['currency']),
    MISSING_LINE_ITEMS: utils_1.omit(exports.SHOPIFY_ORDER_VALUES.valid, ['line_items']),
    NULL_SUPPORTER_UUID: utils_1.override(exports.SHOPIFY_ORDER_VALUES.valid, { supporter_uuid: null }),
};
exports.SHOPIFY_ORDERS = {
    invalid: exports.SHOPIFY_ORDERS_INVALID,
    valid: exports.SHOPIFY_ORDERS_VALID,
};
/**
 * Generates a random Shopify Order.
 *
 * @param overrides
 */
var shopifyOrderGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    var lineItems = utils_1.generateList(utils_1.getInteger(1, 5), ShopifyOrderLineItem_mock_1.lineItemGenerator);
    var lineItemsTotal = lineItems.reduce(function (total, item) { return total + item.price - item.discount; }, 0);
    var donation = utils_1.getRandomElement([0, 1, 5, 10]);
    var postage = utils_1.getRandomElement([1.5, 2.5, 4, 7.5]);
    return __assign({ shopify_id: utils_1.getString(14, { uppercase: false, lowercase: false }), shopify_processed_at: utils_1.getDate().toISOString(), shopify_updated_at: utils_1.getDate().toISOString(), shopify_cancelled_at: utils_1.getRandomElement([null, utils_1.getDate().toISOString()]), shopify_customer_id: utils_1.getString(14, { uppercase: false, lowercase: false }), supporter_uuid: uuid_1.v4(), order_number: "#" + utils_1.getInteger(1000, 999999), cancel_reason: utils_1.getRandomElement(__spreadArray([null], Object.values(ShopifyOrder_1.CancelledReason))), payment_status: utils_1.getRandomElement(Object.values(ShopifyOrder_1.PaymentStatus)), fulfilment_status: utils_1.getRandomElement(Object.values(ShopifyOrder_1.FulfilmentStatus)), postcode: 'SE1 7TP', total_line_items_price: lineItemsTotal, total_discounts: 0.00, subtotal_price: lineItemsTotal, subtotal_includes_vat: validation_1.NumericBool.YES, total_outstanding: 0.00, total_donation_received: donation, total_postage: postage, total_price: lineItemsTotal + donation + postage, total_vat: 0.00, total_refunded: 0.00, currency: validation_1.ErpCurrencies.GBP, line_items: lineItems }, overrides);
};
exports.shopifyOrderGenerator = shopifyOrderGenerator;
