// @comicrelief/data-models/schemas/payments/models/Payin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payin = exports.PayinAudience = void 0;
var yup = __importStar(require("yup"));
var StrictString_1 = require("../../../fields/StrictString");
var validation_1 = require("../../../validation");
var PayinAudience;
(function (PayinAudience) {
    PayinAudience["SCHOOL"] = "school";
    PayinAudience["WORK"] = "work";
    PayinAudience["PUBLIC"] = "public";
    PayinAudience["YOUTH"] = "youth";
})(PayinAudience = exports.PayinAudience || (exports.PayinAudience = {}));
/**
 * Payin Put model
 */
exports.Payin = validation_1.defineModel(validation_1.Namespaces.PAYMENTS, 'Payin', function () { return yup.object({
    id: yup.string().uuid().required(),
    transactionId: yup.string().uuid().required(),
    created: validation_1.TimestampOrNow.required(),
    updated: validation_1.TimestampOrNow.required(),
    address1: StrictString_1.StrictString.required(),
    address2: StrictString_1.StrictString.nullable().default(function () { return null; }).defined(),
    address3: StrictString_1.StrictString.nullable().default(function () { return null; }).defined(),
    town: StrictString_1.StrictString.required(),
    country: StrictString_1.StrictString.required(),
    postcode: StrictString_1.StrictString.required(),
    Event: yup.string().nullable().default(function () { return null; }).defined(),
    audience: yup.string().oneOf(Object.values(PayinAudience)).required(),
    WhatYouDid: yup.string().nullable().default(function () { return null; }).defined(),
    permission: yup.boolean().nullable().default(function () { return null; }).defined(),
    permissionEmail: yup.boolean().nullable().default(function () { return null; }).defined(),
    permissionPhone: yup.boolean().nullable().default(function () { return null; }).defined(),
    permissionPost: yup.boolean().nullable().default(function () { return null; }).defined(),
    permissionSMS: yup.boolean().nullable().default(function () { return null; }).defined(),
    phone: yup.string().nullable().default(function () { return null; }).defined(),
    mobile: yup.string().nullable().default(function () { return null; }).defined(),
    establishmentId: yup.string().nullable().default(function () { return null; }).defined(),
    jobTitle: yup.string().nullable().default(function () { return null; }).defined(),
    schoolType: yup.string().nullable().default(function () { return null; }).defined(),
    establishment_name: yup.string().nullable().default(function () { return null; }).defined(),
    organisation: yup.string().nullable().default(function () { return null; }).defined(),
    organisationFreetext: yup.string().nullable().default(function () { return null; }).defined(),
    fundraising_activity: yup.string().nullable().default(function () { return null; }).defined(),
    ukActiveMember: yup.boolean().nullable().default(function () { return null; }).defined(),
    certificate: yup.boolean().nullable().default(function () { return null; }).defined(),
}); });
