// @comicrelief/test-utils/aws/testers/S3Tester.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3Tester = void 0;
var aws_sdk_1 = __importDefault(require("aws-sdk"));
var Tester_1 = require("../../Tester");
/**
 * Manages cleanup of S3 files
 * during feature tests.
 *
 * @class
 */
var S3Tester = /** @class */ (function (_super) {
    __extends(S3Tester, _super);
    function S3Tester() {
        var _this = _super.call(this) || this;
        _this.client = new aws_sdk_1.default.S3({
            region: process.env.REGION,
            httpOptions: {
                connectTimeout: 2000,
                timeout: 2000,
            },
        });
        return _this;
    }
    /**
     * Performs a HEAD request.
     *
     * @param bucket
     * @param key
     */
    S3Tester.prototype.head = function (bucket, key) {
        return this.client.headObject({ Bucket: bucket, Key: key }).promise();
    };
    /**
     * Lists all objects in the given prefix
     *
     * @param bucket
     * @param prefix
     */
    S3Tester.prototype.list = function (bucket, prefix) {
        if (prefix === void 0) { prefix = ''; }
        return __awaiter(this, void 0, void 0, function () {
            var params, Contents;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        params = {
                            Bucket: bucket,
                            Prefix: prefix,
                        };
                        return [4 /*yield*/, this.client.listObjects(params).promise()];
                    case 1:
                        Contents = (_a.sent()).Contents;
                        return [2 /*return*/, Contents];
                }
            });
        });
    };
    /**
     * @inheritdoc
     */
    S3Tester.prototype.removeReference = function (reference) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.client.deleteObject(reference).promise();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Deletes all objects found
     * in the specific prefix.
     * Useful to clean up the Athena bucket
     * of the query results.
     *
     * @param bucket
     * @param prefix
     */
    S3Tester.prototype.clearPath = function (bucket, prefix) {
        return __awaiter(this, void 0, void 0, function () {
            var results, Objects, params;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.list(bucket, prefix)];
                    case 1:
                        results = _a.sent();
                        if (!results || results.length === 0) {
                            return [2 /*return*/];
                        }
                        Objects = results
                            .filter(function (object) { return Boolean(object.Key); })
                            .map(function (object) { return ({ Key: object.Key }); });
                        params = {
                            Bucket: bucket,
                            Delete: { Objects: Objects },
                        };
                        this.client.deleteObjects(params).promise();
                        return [2 /*return*/];
                }
            });
        });
    };
    return S3Tester;
}(Tester_1.Tester));
exports.S3Tester = S3Tester;
