// @comicrelief/data-models/schemas/prize-platform/mocks/sqs/SQSTicketsEndOfBatchPayload.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSQSTicketsEndOfBatchPayload = exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS = exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_INVALID = exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_VALID = exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES = exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_INVALID_VALUES = exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALID_VALUES = void 0;
var utils_1 = require("../../../../utils");
var SQSTicketsEndOfBatchPayload_1 = require("../../models/sqs/SQSTicketsEndOfBatchPayload");
exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALID_VALUES = {
    event: SQSTicketsEndOfBatchPayload_1.SQSTicketsEndOfBatchEvent.CREATED_POSTAL_TICKETS,
    count: 10,
};
exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_INVALID_VALUES = {
    event: 'invalid event',
    count: -1,
};
exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES = {
    valid: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALID_VALUES,
    invalid: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_INVALID_VALUES,
};
exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_VALID = {
    DEFAULT: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.valid,
};
exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_INVALID = {
    DEFAULT: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.invalid,
    INVALID_EVENT: utils_1.override(exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.valid, { event: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.invalid.event }),
    INVALID_COUNT: utils_1.override(exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.valid, { count: exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.invalid.count }),
    MISSING_EVENT: utils_1.omit(exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.valid, ['event']),
    MISSING_COUNT: utils_1.omit(exports.SQS_TICKETS_ENDOFBATCH_PAYLOAD_VALUES.valid, ['count']),
};
exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS = {
    valid: exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_VALID,
    invalid: exports.SQS_TICKETS_ENDOFBATCH_PAYLOADS_INVALID,
};
/**
 * Generates a random SQSTicketsEndOfBatchPayloadProps
 *
 * @param overrides
 */
var createSQSTicketsEndOfBatchPayload = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ event: utils_1.getRandomElement(Object.values(SQSTicketsEndOfBatchPayload_1.SQSTicketsEndOfBatchEvent)), count: utils_1.getInteger(1, 100) }, overrides));
};
exports.createSQSTicketsEndOfBatchPayload = createSQSTicketsEndOfBatchPayload;
