// @comicrelief/data-models/schemas/prize-platform/mocks/sqs/SQSTicketsVerifyPayload.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSQSTicketsVerifyPayload = exports.SQS_TICKETS_VERIFY_PAYLOADS = exports.SQS_TICKETS_VERIFY_PAYLOADS_INVALID = exports.SQS_TICKETS_VERIFY_PAYLOADS_VALID = exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../../utils");
var Transaction_mock_1 = require("../Transaction.mock");
var SQS_TICKETS_VERIFY_PAYLOAD_VALID_VALUES = {
    transaction_uuid: Transaction_mock_1.TRANSACTION_VALUES.valid.uuid,
};
var SQS_TICKETS_VERIFY_PAYLOAD_INVALID_VALUES = {
    transaction_uuid: Transaction_mock_1.TRANSACTION_VALUES.invalid.uuid,
};
exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES = {
    valid: SQS_TICKETS_VERIFY_PAYLOAD_VALID_VALUES,
    invalid: SQS_TICKETS_VERIFY_PAYLOAD_INVALID_VALUES,
};
exports.SQS_TICKETS_VERIFY_PAYLOADS_VALID = {
    DEFAULT: exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES.valid,
    ANOTHER: utils_1.override(exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES.valid, { transaction_uuid: '420b3256-e6c7-48f7-bd0d-5edfca8a132a' }),
};
exports.SQS_TICKETS_VERIFY_PAYLOADS_INVALID = {
    DEFAULT: exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES.invalid,
    INVALID_TRANSACTION_UUID: utils_1.override(exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES.valid, { transaction_uuid: exports.SQS_TICKETS_VERIFY_PAYLOAD_VALUES.invalid.transaction_uuid }),
};
exports.SQS_TICKETS_VERIFY_PAYLOADS = {
    valid: exports.SQS_TICKETS_VERIFY_PAYLOADS_VALID,
    invalid: exports.SQS_TICKETS_VERIFY_PAYLOADS_INVALID,
};
/**
 * Generates a random SQSTicketsVerifyPayloadProps
 *
 * @param overrides
 */
var createSQSTicketsVerifyPayload = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ transaction_uuid: uuid_1.v4() }, overrides));
};
exports.createSQSTicketsVerifyPayload = createSQSTicketsVerifyPayload;
