// @comicrelief/data-models/schemas/erp/mocks/doctypes/giftaid/HMRCClaim.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hmrcClaimGenerator = exports.HMRC_CLAIM_MOCKS = exports.HMRC_CLAIM_INVALID = exports.HMRC_CLAIM_VALID = exports.HMRC_CLAIM_VALUES = exports.HMRC_CLAIM_VALUES_INVALID = exports.HMRC_CLAIM_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
exports.HMRC_CLAIM_VALUES_VALID = {
    hmrc_reference: 'example-xml-fragment-tba',
    hmrc_date_claimed: '2022-05-01',
    swiftaid_reference: '1df868c3-5b5f-4e8f-90f7-7138fba00ef4',
    submission_id: 'bec25f1e-7d76-4dc1-9b08-be8bd4b2fa63',
};
exports.HMRC_CLAIM_VALUES_INVALID = {
    hmrc_reference: null,
    hmrc_date_claimed: 'not-a-date',
    swiftaid_reference: 'not-a-uuid',
    submission_id: 'not-a-uuid',
};
exports.HMRC_CLAIM_VALUES = {
    valid: exports.HMRC_CLAIM_VALUES_VALID,
    invalid: exports.HMRC_CLAIM_VALUES_INVALID,
};
var _a = utils_1.makeDefaultMocks(exports.HMRC_CLAIM_VALUES.valid, exports.HMRC_CLAIM_VALUES.invalid), valid = _a.valid, invalid = _a.invalid;
exports.HMRC_CLAIM_VALID = __assign({}, valid);
exports.HMRC_CLAIM_INVALID = __assign({}, invalid);
exports.HMRC_CLAIM_MOCKS = {
    valid: exports.HMRC_CLAIM_VALID,
    invalid: exports.HMRC_CLAIM_INVALID,
};
/**
 * Creates a random HMRC Claim
 *
 * @param overrides
 */
var hmrcClaimGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ hmrc_reference: utils_1.getString(10), hmrc_date_claimed: utils_1.getDate().toISOString(), swiftaid_reference: utils_1.getUuid(), submission_id: utils_1.getUuid() }, overrides));
};
exports.hmrcClaimGenerator = hmrcClaimGenerator;
