// @comicrelief/data-models/schemas/user-service/mocks/requests/MatchRequest.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MATCH_REQUESTS = exports.MATCH_REQUESTS_INVALID = exports.MATCH_REQUESTS_VALID = exports.MATCH_REQUEST_VALUES = exports.MATCH_REQUEST_INVALID_VALUES = exports.MATCH_REQUEST_VALID_VALUES = void 0;
var utils_1 = require("../../../../utils");
var SupporterInfo_mock_1 = require("../../../shared/mocks/SupporterInfo.mock");
var User_mock_1 = require("../User.mock");
exports.MATCH_REQUEST_VALID_VALUES = {
    username: User_mock_1.USER_VALUES.valid.uuid,
    email: SupporterInfo_mock_1.EMAILS.valid,
    mobile: SupporterInfo_mock_1.MOBILES.valid.DEFAULT,
    identity_hash: 'abc',
};
exports.MATCH_REQUEST_INVALID_VALUES = {
    username: User_mock_1.USER_VALUES.invalid.uuid,
    email: SupporterInfo_mock_1.EMAILS.invalid,
    mobile: SupporterInfo_mock_1.MOBILES.invalid.DEFAULT,
    identity_hash: null,
};
exports.MATCH_REQUEST_VALUES = {
    valid: exports.MATCH_REQUEST_VALID_VALUES,
    invalid: exports.MATCH_REQUEST_INVALID_VALUES,
};
exports.MATCH_REQUESTS_VALID = {
    DEFAULT: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['email']),
    USERNAME: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['username']),
    EMAIL: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['email']),
    MOBILE: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['mobile']),
    IDENTITY_HASH: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['identity_hash']),
};
exports.MATCH_REQUESTS_INVALID = {
    DEFAULT: exports.MATCH_REQUEST_INVALID_VALUES,
    ALL_VALID: exports.MATCH_REQUEST_VALUES.valid,
    MISSING_ALL: {},
    USERNAME_AND_EMAIL: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['username', 'email']),
    USERNAME_AND_MOBILE: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['username', 'mobile']),
    USERNAME_AND_HASH: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['username', 'identity_hash']),
    EMAIL_AND_MOBILE: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['email', 'mobile']),
    EMAIL_AND_HASH: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['email', 'identity_hash']),
    MOBILE_AND_HASH: utils_1.pick(exports.MATCH_REQUEST_VALUES.valid, ['mobile', 'identity_hash']),
    INVALID_USERNAME: { username: exports.MATCH_REQUEST_INVALID_VALUES.username },
    INVALID_EMAIL: { email: exports.MATCH_REQUEST_INVALID_VALUES.email },
    INVALID_MOBILE: { mobile: exports.MATCH_REQUEST_INVALID_VALUES.mobile },
    INVALID_HASH: { identity_hash: exports.MATCH_REQUEST_INVALID_VALUES.identity_hash },
};
exports.MATCH_REQUESTS = {
    valid: exports.MATCH_REQUESTS_VALID,
    invalid: exports.MATCH_REQUESTS_INVALID,
};
