// @comicrelief/data-models/schemas/user-service/mocks/ErpPayload.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERP_PAYLOADS = exports.ERP_PAYLOADS_INVALID = exports.ERP_PAYLOADS_VALID = exports.ERP_PAYLOAD_VALUES = exports.ERP_PAYLOAD_INVALID_VALUES = exports.ERP_PAYLOAD_VALID_VALUES = exports.ERP_PAYLOAD_IDENTITY_ADD_VALID = exports.ERP_PAYLOAD_USER_CREATE_VALID = void 0;
var utils_1 = require("../../../utils");
var ErpPayload_1 = require("../models/ErpPayload");
var Identity_mock_1 = require("./Identity.mock");
var User_mock_1 = require("./User.mock");
exports.ERP_PAYLOAD_USER_CREATE_VALID = {
    type: ErpPayload_1.ErpPayloadTypes.USER__CREATE,
    cognito_user_id: User_mock_1.USERS.valid.DEFAULT.uuid,
    first_name: User_mock_1.USERS.valid.DEFAULT.firstName,
    last_name: User_mock_1.USERS.valid.DEFAULT.lastName,
    email: [
        {
            id: Identity_mock_1.IDENTITIES.valid.EMAIL.hash,
            email: Identity_mock_1.IDENTITIES.valid.EMAIL.content,
            hidden: Identity_mock_1.IDENTITIES.valid.EMAIL.hidden,
            verified: Identity_mock_1.IDENTITIES.valid.EMAIL.verified,
        },
    ],
    mobile: [
        {
            id: Identity_mock_1.IDENTITIES.valid.MOBILE.hash,
            number: Identity_mock_1.IDENTITIES.valid.MOBILE.content,
            hidden: Identity_mock_1.IDENTITIES.valid.MOBILE.hidden,
            verified: Identity_mock_1.IDENTITIES.valid.MOBILE.verified,
        },
    ],
};
exports.ERP_PAYLOAD_IDENTITY_ADD_VALID = utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { type: ErpPayload_1.ErpPayloadTypes.IDENTITY__ADD });
exports.ERP_PAYLOAD_VALID_VALUES = exports.ERP_PAYLOAD_USER_CREATE_VALID;
exports.ERP_PAYLOAD_INVALID_VALUES = {
    type: 'some-type',
    cognito_user_id: User_mock_1.USERS.invalid.DEFAULT.uuid,
    first_name: User_mock_1.USERS.invalid.DEFAULT.firstName,
    last_name: User_mock_1.USERS.invalid.DEFAULT.lastName,
    email: [{ a: 1 }],
    mobile: [{ b: 2 }],
};
exports.ERP_PAYLOAD_VALUES = {
    valid: exports.ERP_PAYLOAD_VALID_VALUES,
    invalid: exports.ERP_PAYLOAD_INVALID_VALUES,
};
exports.ERP_PAYLOADS_VALID = {
    DEFAULT: exports.ERP_PAYLOAD_VALID_VALUES,
    IDENTITY_ADD: exports.ERP_PAYLOAD_IDENTITY_ADD_VALID,
    IDENTITY_ADD_NO_USER_DETAILS: utils_1.omit(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, ['first_name', 'last_name']),
    IDENTITY_ADD_ONLY_EMAILS: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { mobile: [] }),
    IDENTITY_ADD_ONLY_MOBILES: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { email: [] }),
    IDENTITY_ADD_MULTIPLE_EMAILS: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, {
        email: [exports.ERP_PAYLOAD_IDENTITY_ADD_VALID.email[0], exports.ERP_PAYLOAD_IDENTITY_ADD_VALID.email[0], exports.ERP_PAYLOAD_IDENTITY_ADD_VALID.email[0]],
    }),
    USER_CREATE: exports.ERP_PAYLOAD_USER_CREATE_VALID,
    USER_CREATE_ONLY_EMAILS: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { mobile: [] }),
    USER_CREATE_ONLY_MOBILES: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { email: [] }),
    USER_CREATE_MULTIPLE_EMAILS: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, {
        email: [exports.ERP_PAYLOAD_USER_CREATE_VALID.email[0], exports.ERP_PAYLOAD_USER_CREATE_VALID.email[0], exports.ERP_PAYLOAD_USER_CREATE_VALID.email[0]],
    }),
};
exports.ERP_PAYLOADS_INVALID = {
    DEFAULT: exports.ERP_PAYLOAD_INVALID_VALUES,
    IDENTITY_ADD_MISSING_COGNITO_USER_ID: utils_1.omit(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, ['cognito_user_id']),
    IDENTITY_ADD_MISSING_EMAIL_AND_MOBILE: utils_1.omit(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, ['email', 'mobile']),
    IDENTITY_ADD_INVALID_TYPE: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { type: exports.ERP_PAYLOAD_INVALID_VALUES.type }),
    IDENTITY_ADD_INVALID_COGNITO_USER_ID: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { cognito_user_id: exports.ERP_PAYLOAD_INVALID_VALUES.cognito_user_id }),
    IDENTITY_ADD_INVALID_FIRST_NAME: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { first_name: exports.ERP_PAYLOAD_INVALID_VALUES.first_name }),
    IDENTITY_ADD_INVALID_LAST_NAME: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { last_name: exports.ERP_PAYLOAD_INVALID_VALUES.last_name }),
    IDENTITY_ADD_INVALID_EMAIL: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { email: exports.ERP_PAYLOAD_INVALID_VALUES.email }),
    IDENTITY_ADD_INVALID_MOBILE: utils_1.override(exports.ERP_PAYLOAD_IDENTITY_ADD_VALID, { mobile: exports.ERP_PAYLOAD_INVALID_VALUES.mobile }),
    USER_CREATE_MISSING_COGNITO_USER_ID: utils_1.omit(exports.ERP_PAYLOAD_USER_CREATE_VALID, ['cognito_user_id']),
    USER_CREATE_MISSING_FIRST_NAME: utils_1.omit(exports.ERP_PAYLOAD_USER_CREATE_VALID, ['first_name']),
    USER_CREATE_MISSING_LAST_NAME: utils_1.omit(exports.ERP_PAYLOAD_USER_CREATE_VALID, ['last_name']),
    USER_CREATE_MISSING_EMAIL_AND_MOBILE: utils_1.omit(exports.ERP_PAYLOAD_USER_CREATE_VALID, ['email', 'mobile']),
    USER_CREATE_INVALID_TYPE: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { type: exports.ERP_PAYLOAD_INVALID_VALUES.type }),
    USER_CREATE_INVALID_COGNITO_USER_ID: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { cognito_user_id: exports.ERP_PAYLOAD_INVALID_VALUES.cognito_user_id }),
    USER_CREATE_INVALID_FIRST_NAME: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { first_name: exports.ERP_PAYLOAD_INVALID_VALUES.first_name }),
    USER_CREATE_INVALID_LAST_NAME: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { last_name: exports.ERP_PAYLOAD_INVALID_VALUES.last_name }),
    USER_CREATE_INVALID_EMAIL: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { email: exports.ERP_PAYLOAD_INVALID_VALUES.email }),
    USER_CREATE_INVALID_MOBILE: utils_1.override(exports.ERP_PAYLOAD_USER_CREATE_VALID, { mobile: exports.ERP_PAYLOAD_INVALID_VALUES.mobile }),
};
exports.ERP_PAYLOADS = {
    valid: exports.ERP_PAYLOADS_VALID,
    invalid: exports.ERP_PAYLOADS_INVALID,
};
