// @comicrelief/data-models/schemas/user-service/mocks/User.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.userGenerator = exports.USERS = exports.USERS_INVALID = exports.USERS_VALID = exports.USER_VALUES = exports.USER_INVALID_VALUES = exports.USER_VALID_VALUES = void 0;
var uuid_1 = require("uuid");
var utils_1 = require("../../../utils");
exports.USER_VALID_VALUES = {
    uuid: 'b6478c65-3109-447a-ace8-9e8ecdd8f590',
    firstName: 'Dante',
    lastName: 'Alighieri',
    active: true,
    created: new Date(1599748412553),
    updated: new Date(1599748412553),
};
exports.USER_INVALID_VALUES = {
    uuid: 'invalid-uuid',
    firstName: null,
    lastName: null,
    active: Number.NaN,
    created: 'not-a-created-date',
    updated: 'not-a-updated-date',
};
exports.USER_VALUES = {
    valid: exports.USER_VALID_VALUES,
    invalid: exports.USER_INVALID_VALUES,
};
exports.USERS_VALID = {
    DEFAULT: exports.USER_VALID_VALUES,
    MISSING_UUID: utils_1.omit(exports.USER_VALID_VALUES, ['uuid']),
    MISSING_ACTIVE: utils_1.omit(exports.USER_VALID_VALUES, ['active']),
    MISSING_CREATED: utils_1.omit(exports.USER_VALID_VALUES, ['created']),
    MISSING_UPDATED: utils_1.omit(exports.USER_VALID_VALUES, ['updated']),
    MISSING_FIRSTNAME: utils_1.omit(exports.USER_VALID_VALUES, ['firstName']),
    MISSING_LASTNAME: utils_1.omit(exports.USER_VALID_VALUES, ['lastName']),
    DOUBLE_FIRSTNAME: utils_1.override(exports.USER_VALID_VALUES, { firstName: 'Dante Poeta' }),
    DOUBLE_SURNAME: utils_1.override(exports.USER_VALID_VALUES, { lastName: 'Alighieri Poeta' }),
};
exports.USERS_INVALID = {
    DEFAULT: exports.USER_INVALID_VALUES,
    INVALID_UUID: utils_1.override(exports.USER_VALID_VALUES, { uuid: exports.USER_INVALID_VALUES.uuid }),
    INVALID_FIRSTNAME: utils_1.override(exports.USER_VALID_VALUES, { firstName: exports.USER_INVALID_VALUES.firstName }),
    INVALID_LASTNAME: utils_1.override(exports.USER_VALID_VALUES, { lastName: exports.USER_INVALID_VALUES.lastName }),
    INVALID_CREATED: utils_1.override(exports.USER_VALID_VALUES, { created: exports.USER_INVALID_VALUES.created }),
    INVALID_UPDATED: utils_1.override(exports.USER_VALID_VALUES, { updated: exports.USER_INVALID_VALUES.updated }),
};
exports.USERS = {
    valid: exports.USERS_VALID,
    invalid: exports.USERS_INVALID,
};
/**
 * Generates a random user
 *
 * @param overrides
 */
var userGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    // One random time in the last week
    var timestamp = utils_1.getUnixTimestamp();
    return __assign({ uuid: uuid_1.v4(), firstName: utils_1.getString(10, { numbers: false }), lastName: utils_1.getString(16, { numbers: false }), active: false, created: new Date(timestamp), updated: new Date(timestamp) }, overrides);
};
exports.userGenerator = userGenerator;
