// @comicrelief/data-models/schemas/prize-platform/models/Config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.ConfigStates = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../validation");
/**
 * Represents the service states
 */
var ConfigStates;
(function (ConfigStates) {
    ConfigStates["OK"] = "OK";
    ConfigStates["TEMPORARILY_PAUSED"] = "TEMPORARILY_PAUSED";
    ConfigStates["INDEFINITELY_PAUSED"] = "INDEFINITELY_PAUSED";
})(ConfigStates = exports.ConfigStates || (exports.ConfigStates = {}));
/**
 * Config model.
 */
exports.Config = validation_1.defineModel(validation_1.Namespaces.PRIZE_PLATFORM_SERVICE, 'Config', function () { return yup
    .object({
    state: yup.string().oneOf(Object.values(ConfigStates)).required(),
}); });
