// @comicrelief/data-models/schemas/prize-platform/mocks/requests/notify/ParticipantsRequest.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.participantsRequestGenerator = exports.PARTICIPANTS_REQUESTS = exports.PARTICIPANTS_REQUESTS_INVALID = exports.PARTICIPANTS_REQUESTS_VALID = exports.PARTICIPANTS_REQUEST_VALUES = exports.PARTICIPANTS_REQUEST_INVALID_VALUES = exports.PARTICIPANTS_REQUEST_VALID_VALUES = void 0;
var utils_1 = require("../../../../../utils");
var Transaction_mock_1 = require("../../Transaction.mock");
exports.PARTICIPANTS_REQUEST_VALID_VALUES = {
    prize_id: Transaction_mock_1.TRANSACTION_VALUES.valid.prize_id,
};
exports.PARTICIPANTS_REQUEST_INVALID_VALUES = {
    prize_id: Transaction_mock_1.TRANSACTION_VALUES.invalid.prize_id,
};
exports.PARTICIPANTS_REQUEST_VALUES = {
    valid: exports.PARTICIPANTS_REQUEST_VALID_VALUES,
    invalid: exports.PARTICIPANTS_REQUEST_INVALID_VALUES,
};
exports.PARTICIPANTS_REQUESTS_VALID = {
    DEFAULT: exports.PARTICIPANTS_REQUEST_VALUES.valid,
};
exports.PARTICIPANTS_REQUESTS_INVALID = {
    DEFAULT: exports.PARTICIPANTS_REQUEST_VALUES.invalid,
    INVALID_PRIZE_ID: utils_1.override(exports.PARTICIPANTS_REQUEST_VALUES.valid, { prize_id: exports.PARTICIPANTS_REQUEST_VALUES.invalid.prize_id }),
    MISSING_PRIZE_ID: utils_1.omit(exports.PARTICIPANTS_REQUEST_VALUES.valid, ['prize_id']),
};
exports.PARTICIPANTS_REQUESTS = {
    valid: exports.PARTICIPANTS_REQUESTS_VALID,
    invalid: exports.PARTICIPANTS_REQUESTS_INVALID,
};
/**
 * Generates a random ParticipantsRequest
 *
 * @param overrides
 */
var participantsRequestGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ prize_id: utils_1.getString(10) }, overrides));
};
exports.participantsRequestGenerator = participantsRequestGenerator;
