// @comicrelief/data-models/schemas/erp/mocks/actions/psl/PayinPut.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePayinPut = exports.PAYIN_PUTS = exports.PAYIN_PUTS_INVALID = exports.PAYIN_PUTS_VALID = exports.PAYIN_PUT_VALUES = exports.PAYIN_PUT_VALUES_INVALID = exports.PAYIN_PUT_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
var Payin_mock_1 = require("../../../../payments/mocks/Payin.mock");
var common_1 = require("../../../models/common");
exports.PAYIN_PUT_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.PAYIN__PUT,
    data: __assign({}, Payin_mock_1.PAYINS.valid.DEFAULT),
};
exports.PAYIN_PUT_VALUES_INVALID = {
    type: 'some-type',
    data: __assign({}, Payin_mock_1.PAYINS.invalid.DEFAULT),
};
exports.PAYIN_PUT_VALUES = {
    valid: exports.PAYIN_PUT_VALUES_VALID,
    invalid: exports.PAYIN_PUT_VALUES_INVALID,
};
exports.PAYIN_PUTS_VALID = __assign({ DEFAULT: exports.PAYIN_PUT_VALUES.valid }, utils_1.overrideFromMocks(exports.PAYIN_PUT_VALUES.valid, 'data', Payin_mock_1.PAYINS.valid));
exports.PAYIN_PUTS_INVALID = __assign({ DEFAULT: utils_1.override(exports.PAYIN_PUT_VALUES_VALID, { data: exports.PAYIN_PUT_VALUES.invalid.data }), INVALID_TYPE: utils_1.override(exports.PAYIN_PUT_VALUES_VALID, { type: exports.PAYIN_PUT_VALUES.invalid.type }), INVALID_DATA: utils_1.override(exports.PAYIN_PUT_VALUES_VALID, { data: null }), MISSING_TYPE: utils_1.omit(exports.PAYIN_PUT_VALUES_VALID, ['type']), MISSING_DATA: utils_1.omit(exports.PAYIN_PUT_VALUES_VALID, ['data']) }, utils_1.overrideFromMocks(exports.PAYIN_PUT_VALUES.valid, 'data', Payin_mock_1.PAYINS.invalid));
exports.PAYIN_PUTS = {
    invalid: exports.PAYIN_PUTS_INVALID,
    valid: exports.PAYIN_PUTS_VALID,
};
/**
 * Generates a random PostalTicketPut
 *
 * @param overrides
 */
var generatePayinPut = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.PAYIN__PUT, data: __assign({}, Payin_mock_1.payinGenerator()) }, overrides));
};
exports.generatePayinPut = generatePayinPut;
