// @comicrelief/data-models/schemas/erp/models/doctypes/giftaid/GiftAidDeclaration.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GiftAidDeclaration = exports.GiftAidSource = exports.DOCTYPE = void 0;
var yup = __importStar(require("yup"));
var validation_1 = require("../../../../../validation");
var common_1 = require("../../common");
/**
 * Name of this doctype in ERP.
 */
exports.DOCTYPE = common_1.ErpResources.GIFTAID_DECLARATION;
/**
 * Represents the source of the Gift Aid declaration
 */
var GiftAidSource;
(function (GiftAidSource) {
    GiftAidSource["GIFTAID_SUBMIT"] = "Giftaid Submit";
    GiftAidSource["GIFTAID_UPDATE"] = "Giftaid Update";
    GiftAidSource["GIFTAID_SUBMIT_IVR"] = "Giftaid Submit IVR";
})(GiftAidSource = exports.GiftAidSource || (exports.GiftAidSource = {}));
exports.GiftAidDeclaration = validation_1.defineModel(validation_1.Namespaces.ERP_SERVICE, exports.DOCTYPE, function () { return yup.object({
    declaration_transaction_id: yup.string().uuid().required(),
    supporter_uuid: yup.string().uuid().required(),
    declaration_start_date: validation_1.Timestamp.required(),
    declaration_end_date: validation_1.Timestamp.nullable().default(function () { return null; }).defined(),
    submission_date: validation_1.Timestamp.required(),
    first_name: yup.string().required(),
    last_name: yup.string().required(),
    source: yup.string().oneOf(Object.values(GiftAidSource)).required(),
    can_claim_giftaid: validation_1.NumericBoolField.required(),
    postal_id: yup.string().required(),
}); });
