// @comicrelief/data-models/schemas/marketing-preferences/mocks/requests/Submit.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.submitRequestGenerator = exports.SUBMIT_REQUESTS = exports.SUBMIT_REQUESTS_INVALID = exports.SUBMIT_REQUESTS_VALID = exports.SUBMIT_REQUEST_VALUES = exports.VALUES_INVALID = exports.VALUES_VALID = void 0;
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
var transsource_1 = require("../../transsource");
exports.VALUES_VALID = {
    firstname: 'Dante',
    lastname: 'Alighieri',
    mobile: '+447890123456',
    phone: '+447890123456',
    email: 'dante@example.com',
    address1: 'Camelford House',
    address2: '89 Albert Embankment',
    address3: 'Vauxhall',
    town: 'London',
    postcode: 'SE1 7TP',
    country: 'United Kingdom',
    campaign: enums_1.Campaigns.RND21,
    transSource: transsource_1.sources.REACT_MARKETING_PREFERENCES,
    transSourceUrl: 'https://donate.comicrelief.com',
    transType: 'prefs',
    permissionEmail: 1,
    permissionPhone: 0,
    permissionPost: 0,
    permissionSMS: 1,
    transactionId: '5f1aaa96-54d2-4a55-8fdf-a7f20291dc1a',
};
exports.VALUES_INVALID = {
    firstname: null,
    lastname: null,
    mobile: 12345,
    phone: 12345,
    email: 'not-an-email-address',
    address1: '32<%',
    address2: 0,
    address3: 0,
    town: 0,
    postcode: 'not-a-postcode',
    country: 0,
    campaign: 'not-a-campaign',
    transSource: null,
    transSourceUrl: 'not-a-url',
    transType: 'not-prefs',
    permissionEmail: 'not-a-permission',
    permissionPhone: 'not-a-permission',
    permissionPost: 'not-a-permission',
    permissionSMS: 'not-a-permission',
    transactionId: 'not-a-uuid',
};
exports.SUBMIT_REQUEST_VALUES = {
    valid: exports.VALUES_VALID,
    invalid: exports.VALUES_INVALID,
};
/* eslint-disable object-curly-newline */
exports.SUBMIT_REQUESTS_VALID = {
    DEFAULT: exports.SUBMIT_REQUEST_VALUES.valid,
    // optional field (check in the snapshot that it defaults to null)
    MISSING_TRANSACTION_ID: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['transactionId']),
    NULL_TRANSACTION_ID: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transactionId: null }),
    // these are allowed if the corresponding permission is missing/null
    MISSING_ADDRESS: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionPost', 'address1', 'address2', 'address3', 'town', 'postcode', 'country']),
    MISSING_EMAIL: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionEmail', 'email']),
    MISSING_PHONE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionPhone', 'phone']),
    MISSING_SMS: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionSMS', 'mobile']),
    NULL_ADDRESS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPost: null, address1: null, address2: null, address3: null, town: null, postcode: null, country: null }),
    NULL_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionEmail: null, email: null }),
    NULL_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPhone: null, phone: null }),
    NULL_SMS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionSMS: null, mobile: null }),
    EMPTY_ADDRESS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPost: null, address1: '', address2: '', address3: '', town: '', postcode: '', country: '' }),
    EMPTY_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionEmail: null, email: '' }),
    EMPTY_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPhone: null, phone: '' }),
    EMPTY_SMS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionSMS: null, mobile: '' }),
    // address lines 2 and 3 are always optional
    MISSING_ADDRESS2: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['address2']),
    MISSING_ADDRESS2_AND_3: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['address2', 'address3']),
    NULL_ADDRESS2: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address2: null }),
    NULL_ADDRESS2_AND_3: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address2: null, address3: null }),
    EMPTY_ADDRESS2: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address2: '' }),
    EMPTY_ADDRESS2_AND_3: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address2: '', address3: '' }),
    // permission fields are optional (provided there is at least one)
    MISSING_PERMISSIONEMAIL: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionEmail']),
    MISSING_PERMISSIONPHONE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionPhone']),
    MISSING_PERMISSIONPOST: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionPost']),
    MISSING_PERMISSIONSMS: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionSMS']),
    // transformed values
    LOWERCASE_CAMPAIGN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { campaign: 'rnd22' }),
    LOWERCASE_POSTCODE_WITHOUT_SPACE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { postcode: 'se17tp' }),
};
exports.SUBMIT_REQUESTS_INVALID = {
    DEFAULT: exports.SUBMIT_REQUEST_VALUES.invalid,
    INVALID_FIRSTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { firstname: exports.VALUES_INVALID.firstname }),
    INVALID_LASTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { lastname: exports.VALUES_INVALID.lastname }),
    INVALID_MOBILE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { mobile: exports.VALUES_INVALID.mobile }),
    INVALID_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { phone: exports.VALUES_INVALID.phone }),
    INVALID_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { email: exports.VALUES_INVALID.email }),
    INVALID_ADDRESS1: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address1: exports.VALUES_INVALID.address1 }),
    INVALID_ADDRESS2: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address2: exports.VALUES_INVALID.address2 }),
    INVALID_ADDRESS3: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address3: exports.VALUES_INVALID.address3 }),
    INVALID_TOWN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { town: exports.VALUES_INVALID.town }),
    INVALID_POSTCODE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { postcode: exports.VALUES_INVALID.postcode }),
    INVALID_COUNTRY: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { country: exports.VALUES_INVALID.country }),
    INVALID_CAMPAIGN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { campaign: exports.VALUES_INVALID.campaign }),
    INVALID_TRANSSOURCE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSource: exports.VALUES_INVALID.transSource }),
    INVALID_TRANSSOURCEURL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSourceUrl: exports.VALUES_INVALID.transSourceUrl }),
    INVALID_TRANSTYPE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transType: exports.VALUES_INVALID.transType }),
    INVALID_PERMISSIONEMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionEmail: exports.VALUES_INVALID.permissionEmail }),
    INVALID_PERMISSIONPHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPhone: exports.VALUES_INVALID.permissionPhone }),
    INVALID_PERMISSIONPOST: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionPost: exports.VALUES_INVALID.permissionPost }),
    INVALID_PERMISSIONSMS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionSMS: exports.VALUES_INVALID.permissionSMS }),
    INVALID_TRANSACTIONID: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transactionId: exports.VALUES_INVALID.transactionId }),
    // these are required fields
    MISSING_FIRSTNAME: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['firstname']),
    MISSING_LASTNAME: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['lastname']),
    MISSING_CAMPAIGN: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['campaign']),
    MISSING_TRANSSOURCE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['transSource']),
    MISSING_TRANSSOURCEURL: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['transSourceUrl']),
    MISSING_TRANSTYPE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['transType']),
    EMPTY_FIRSTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { firstname: '' }),
    EMPTY_LASTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { lastname: '' }),
    EMPTY_CAMPAIGN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { campaign: '' }),
    EMPTY_TRANSSOURCE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSource: '' }),
    EMPTY_TRANSSOURCEURL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSourceUrl: '' }),
    EMPTY_TRANSTYPE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transType: '' }),
    NULL_FIRSTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { firstname: null }),
    NULL_LASTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { lastname: null }),
    NULL_CAMPAIGN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { campaign: null }),
    NULL_TRANSSOURCE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSource: null }),
    NULL_TRANSSOURCEURL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transSourceUrl: null }),
    NULL_TRANSTYPE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { transType: null }),
    BLANK_FIRSTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { firstname: ' ' }),
    BLANK_LASTNAME: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { lastname: ' ' }),
    // at least one permission must be specified
    MISSING_ALL_PERMISSIONS: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['permissionEmail', 'permissionPhone', 'permissionPost', 'permissionSMS']),
    NULL_ALL_PERMISSIONS: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { permissionEmail: null, permissionPhone: null, permissionPost: null, permissionSMS: null }),
    // these are invalid because `permissionPost` is non-null
    MISSING_ADDRESS1: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['address1']),
    MISSING_TOWN: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['town']),
    MISSING_POSTCODE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['postcode']),
    MISSING_COUNTRY: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['country']),
    NULL_ADDRESS1: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address1: null }),
    NULL_TOWN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { town: null }),
    NULL_POSTCODE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { postcode: null }),
    NULL_COUNTRY: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { country: null }),
    EMPTY_ADDRESS1: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address1: '' }),
    EMPTY_TOWN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { town: '' }),
    EMPTY_POSTCODE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { postcode: '' }),
    EMPTY_COUNTRY: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { country: '' }),
    BLANK_ADDRESS1: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { address1: ' ' }),
    BLANK_TOWN: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { town: ' ' }),
    BLANK_POSTCODE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { postcode: ' ' }),
    BLANK_COUNTRY: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { country: ' ' }),
    // invalid because `permissionEmail` is non-null
    MISSING_EMAIL: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['email']),
    NULL_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { email: null }),
    EMPTY_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { email: '' }),
    BLANK_EMAIL: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { email: ' ' }),
    // invalid because `permissionPhone` is non-null
    MISSING_PHONE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['phone']),
    NULL_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { phone: null }),
    EMPTY_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { phone: '' }),
    BLANK_PHONE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { phone: ' ' }),
    // invalid because `permissionSMS` is non-null
    MISSING_MOBILE: utils_1.omit(exports.SUBMIT_REQUEST_VALUES.valid, ['mobile']),
    NULL_MOBILE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { mobile: null }),
    EMPTY_MOBILE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { mobile: '' }),
    BLANK_MOBILE: utils_1.override(exports.SUBMIT_REQUEST_VALUES.valid, { mobile: ' ' }),
};
exports.SUBMIT_REQUESTS = {
    valid: exports.SUBMIT_REQUESTS_VALID,
    invalid: exports.SUBMIT_REQUESTS_INVALID,
};
/**
 * Generates a random fundraiser signup Parquet record.
 *
 * @param overrides
 */
var submitRequestGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ firstname: utils_1.getString(10), lastname: utils_1.getString(10), mobile: utils_1.getMobile(), phone: utils_1.getMobile(), email: utils_1.getEmail(), address1: utils_1.getString(10), address2: utils_1.getString(10), address3: utils_1.getString(10), town: utils_1.getString(10), postcode: exports.VALUES_VALID.postcode, country: exports.VALUES_VALID.country, campaign: utils_1.getRandomElement(enums_1.Campaigns), transSource: utils_1.getString(10), transSourceUrl: utils_1.getUrl(), transType: 'prefs', permissionEmail: utils_1.getRandomElement([0, 1, null]), permissionPhone: utils_1.getRandomElement([0, 1, null]), permissionPost: utils_1.getRandomElement([0, 1, null]), permissionSMS: utils_1.getRandomElement([0, 1, null]) }, overrides));
};
exports.submitRequestGenerator = submitRequestGenerator;
