// @comicrelief/data-models/schemas/erp/mocks/doctypes/FundraisingSegment.mock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fundraisingSegmentGenerator = exports.FUNDRAISING_SEGMENT_MOCKS = exports.FUNDRAISING_SEGMENT_INVALID = exports.FUNDRAISING_SEGMENT_VALID = exports.FUNDRAISING_SEGMENT_VALUES = exports.FUNDRAISING_SEGMENT_VALUES_INVALID = exports.FUNDRAISING_SEGMENT_VALUES_VALID = void 0;
var lodash_1 = __importDefault(require("lodash"));
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
exports.FUNDRAISING_SEGMENT_VALUES_VALID = {
    code: 'secondary_school',
    label: 'Secondary School',
    fundraising_group: enums_1.FundraisingGroups.SCHOOL,
};
exports.FUNDRAISING_SEGMENT_VALUES_INVALID = {
    code: 'Secondary_school',
    label: 12345,
    fundraising_group: 'invalid',
};
exports.FUNDRAISING_SEGMENT_VALUES = {
    valid: exports.FUNDRAISING_SEGMENT_VALUES_VALID,
    invalid: exports.FUNDRAISING_SEGMENT_VALUES_INVALID,
};
exports.FUNDRAISING_SEGMENT_VALID = {
    DEFAULT: exports.FUNDRAISING_SEGMENT_VALUES.valid,
};
exports.FUNDRAISING_SEGMENT_INVALID = {
    DEFAULT: exports.FUNDRAISING_SEGMENT_VALUES.invalid,
    MISSING_CODE: utils_1.omit(exports.FUNDRAISING_SEGMENT_VALUES.valid, ['code']),
    MISSING_LABEL: utils_1.omit(exports.FUNDRAISING_SEGMENT_VALUES.valid, ['label']),
    MISSING_FUNDRAISING_GROUP: utils_1.omit(exports.FUNDRAISING_SEGMENT_VALUES.valid, ['fundraising_group']),
    INVALID_CODE: utils_1.override(exports.FUNDRAISING_SEGMENT_VALUES.valid, { code: exports.FUNDRAISING_SEGMENT_VALUES.invalid.code }),
    INVALID_LABEL: utils_1.override(exports.FUNDRAISING_SEGMENT_VALUES.valid, { label: exports.FUNDRAISING_SEGMENT_VALUES.invalid.label }),
    INVALID_FUNDRAISING_GROUP: utils_1.override(exports.FUNDRAISING_SEGMENT_VALUES.valid, { fundraising_group: exports.FUNDRAISING_SEGMENT_VALUES.invalid.fundraising_group }),
};
exports.FUNDRAISING_SEGMENT_MOCKS = {
    valid: exports.FUNDRAISING_SEGMENT_VALID,
    invalid: exports.FUNDRAISING_SEGMENT_INVALID,
};
/**
 * Creates a random FundraisingSegment
 *
 * @param overrides
 */
var fundraisingSegmentGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    var label = overrides.label || utils_1.getString(10);
    return {
        label: label,
        code: overrides.code || lodash_1.default.snakeCase(label),
        fundraising_group: overrides.fundraising_group || utils_1.getRandomElement(Object.values(enums_1.FundraisingGroups)),
    };
};
exports.fundraisingSegmentGenerator = fundraisingSegmentGenerator;
