// @comicrelief/data-models/schemas/erp/mocks/doctypes/FundraisingTool.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fundraisingToolGenerator = exports.FUNDRAISING_TOOL_MOCKS = exports.FUNDRAISING_TOOL_INVALID = exports.FUNDRAISING_TOOL_VALID = exports.FUNDRAISING_TOOL_VALUES = exports.FUNDRAISING_TOOL_VALUES_INVALID = exports.FUNDRAISING_TOOL_VALUES_VALID = void 0;
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
exports.FUNDRAISING_TOOL_VALUES_VALID = {
    code: enums_1.FundraisingTools.JUSTGIVING_PAGE,
    label: 'Set up JustGiving page',
};
exports.FUNDRAISING_TOOL_VALUES_INVALID = {
    code: 'invalid',
    label: { invalid: true },
};
exports.FUNDRAISING_TOOL_VALUES = {
    valid: exports.FUNDRAISING_TOOL_VALUES_VALID,
    invalid: exports.FUNDRAISING_TOOL_VALUES_INVALID,
};
var _a = utils_1.makeDefaultMocks(exports.FUNDRAISING_TOOL_VALUES.valid, exports.FUNDRAISING_TOOL_VALUES.invalid), valid = _a.valid, invalid = _a.invalid;
exports.FUNDRAISING_TOOL_VALID = __assign({}, valid);
exports.FUNDRAISING_TOOL_INVALID = __assign({}, invalid);
exports.FUNDRAISING_TOOL_MOCKS = {
    valid: exports.FUNDRAISING_TOOL_VALID,
    invalid: exports.FUNDRAISING_TOOL_INVALID,
};
/**
 * Creates a random FundraisingTool
 *
 * @param overrides
 */
var fundraisingToolGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ code: utils_1.getRandomElement(Object.values(enums_1.FundraisingTools)), label: utils_1.getString(10) }, overrides));
};
exports.fundraisingToolGenerator = fundraisingToolGenerator;
