// @comicrelief/data-models/schemas/erp/mocks/doctypes/FundraiserSignup.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fundraiserSignupGenerator = exports.FUNDRAISER_SIGNUP_MOCKS = exports.FUNDRAISER_SIGNUP_INVALID = exports.FUNDRAISER_SIGNUP_VALID = void 0;
var uuid_1 = require("uuid");
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
var FUNDRAISER_SIGNUP_VALUES_VALID = {
    uuid: '7f9490eb-9645-44e6-8fa5-8a0d8eee7ba3',
    supporter_uuid: 'f23aad50-fe73-46fc-b7fd-1c68c13e5fe6',
    organisation_uuid: '4497761d-3086-474e-98ac-a58833a46eeb',
    postal_id: '2746aee7-fad6-4670-a821-c867093719d1',
    campaign_code: enums_1.Campaigns.RND21,
    fundraising_segment: enums_1.FundraisingSegmentsSchool.SECONDARY,
    target: '£100 - £500',
    phone: '020 7820 2000',
    justgiving_page_id: 12345,
    submitted: '2021-02-08T08:31:02.160Z',
    fundraising_tools: [],
    fundraising_ideas: [],
};
var FUNDRAISER_SIGNUP_VALUES_INVALID = {
    uuid: 'not-a-uuid',
    supporter_uuid: 'not-a-uuid',
    organisation_uuid: 'not-a-uuid',
    postal_id: null,
    campaign_code: 'unknown code',
    fundraising_segment: false,
    phone: '1234',
    target: 100,
    justgiving_page_id: '12345',
    submitted: 'not-a-date',
    fundraising_tools: 'not-an-array',
    fundraising_ideas: 'not-an-array',
};
var FUNDRAISER_SIGNUP_VALUES = {
    valid: FUNDRAISER_SIGNUP_VALUES_VALID,
    invalid: FUNDRAISER_SIGNUP_VALUES_INVALID,
};
var _a = utils_1.makeDefaultMocks(FUNDRAISER_SIGNUP_VALUES.valid, FUNDRAISER_SIGNUP_VALUES.invalid, 
// nullable
[
    'organisation_uuid',
    'justgiving_page_id',
    'target',
    'phone',
], 
// optional
[
    'target',
    'fundraising_tools',
    'fundraising_ideas',
    'phone',
]), valid = _a.valid, invalid = _a.invalid;
exports.FUNDRAISER_SIGNUP_VALID = __assign({}, valid);
exports.FUNDRAISER_SIGNUP_INVALID = __assign({}, invalid);
exports.FUNDRAISER_SIGNUP_MOCKS = {
    valid: exports.FUNDRAISER_SIGNUP_VALID,
    invalid: exports.FUNDRAISER_SIGNUP_INVALID,
};
/**
 * Generates a random fundraiser signup.
 *
 * @param overrides
 */
var fundraiserSignupGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ uuid: uuid_1.v4(), supporter_uuid: uuid_1.v4(), organisation_uuid: utils_1.getRandomElement([uuid_1.v4(), null]), postal_id: utils_1.getString(10), campaign_code: utils_1.getRandomElement(Object.values(enums_1.Campaigns)), fundraising_segment: utils_1.getRandomElement(Object.values(enums_1.FundraisingSegments)), target: utils_1.getRandomElement(['', utils_1.getString(4)]), justgiving_page_id: utils_1.getRandomElement([utils_1.getInteger(1000, 9999), null]), submitted: utils_1.getTimestamp() }, overrides));
};
exports.fundraiserSignupGenerator = fundraiserSignupGenerator;
