// @comicrelief/data-models/schemas/erp/mocks/doctypes/SupporterOrganisation.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.supporterOrganisationGenerator = exports.SUPPORTER_ORGANISATION_MOCKS = exports.SUPPORTER_ORGANISATION_INVALID = exports.SUPPORTER_ORGANISATION_VALID = exports.SUPPORTER_ORGANISATION_VALUES = exports.SUPPORTER_ORGANISATION_VALUES_INVALID = exports.SUPPORTER_ORGANISATION_VALUES_VALID = void 0;
var enums_1 = require("../../../../enums");
var utils_1 = require("../../../../utils");
exports.SUPPORTER_ORGANISATION_VALUES_VALID = {
    uuid: '0ab44d09-0688-4a65-ba9c-37645a08313c',
    supporter_uuid: '174f8f8b-da5b-468d-9ff1-10ab6842571d',
    organisation_uuid: '0f35ce16-81b9-48a9-b86b-c9272d85943b',
    job_title: 'Head Teacher',
    start: '2020-03-12T12:37:58.147Z',
    end: '2021-03-12T12:37:58.147Z',
    source: enums_1.Sources.FUNDRAISER_SIGNUP,
    source_id: 'bc77a919-3d1e-49ac-a1c1-9747509dc2e7',
};
exports.SUPPORTER_ORGANISATION_VALUES_INVALID = {
    uuid: 'not-a-uuid',
    supporter_uuid: 'not-a-uuid',
    organisation_uuid: 'not-a-uuid',
    job_title: ['not-a-string'],
    start: 'not-a-datetime',
    end: 'not-a-datetime',
    source: 'invalid_source',
    source_id: { key: 'not-a-string' },
};
exports.SUPPORTER_ORGANISATION_VALUES = {
    valid: exports.SUPPORTER_ORGANISATION_VALUES_VALID,
    invalid: exports.SUPPORTER_ORGANISATION_VALUES_INVALID,
};
var _a = utils_1.makeDefaultMocks(exports.SUPPORTER_ORGANISATION_VALUES.valid, exports.SUPPORTER_ORGANISATION_VALUES.invalid, [
    'job_title',
], [
    'uuid',
    'job_title',
    'start',
    'end',
    'source_id',
]), valid = _a.valid, invalid = _a.invalid;
exports.SUPPORTER_ORGANISATION_VALID = __assign({}, valid);
exports.SUPPORTER_ORGANISATION_INVALID = __assign({}, invalid);
exports.SUPPORTER_ORGANISATION_MOCKS = {
    valid: exports.SUPPORTER_ORGANISATION_VALID,
    invalid: exports.SUPPORTER_ORGANISATION_INVALID,
};
/**
 * Creates a random SupporterOrganisation
 *
 * @param overrides
 */
var supporterOrganisationGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ uuid: utils_1.getUuid(), supporter_uuid: utils_1.getUuid(), organisation_uuid: utils_1.getUuid(), job_title: utils_1.getRandomElement([utils_1.getString(10), '']), start: utils_1.getRandomElement([utils_1.getTimestamp(), '']), end: utils_1.getRandomElement([utils_1.getTimestamp(), '']), source: utils_1.getRandomElement(Object.values(enums_1.Sources)), source_id: utils_1.getRandomElement([utils_1.getUuid(), '']) }, overrides));
};
exports.supporterOrganisationGenerator = supporterOrganisationGenerator;
