// @comicrelief/data-models/schemas/erp/mocks/actions/prize/PostalTicketSetState.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPostalTicketSetState = exports.POSTAL_TICKET_SETSTATES = exports.POSTAL_TICKET_SETSTATES_INVALID = exports.POSTAL_TICKET_SETSTATES_VALID = exports.POSTAL_TICKET_SETSTATE_VALUES = exports.POSTAL_TICKET_SETSTATE_VALUES_INVALID = exports.POSTAL_TICKET_SETSTATE_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
var PostalTicket_1 = require("../../../../prize-platform/models/PostalTicket");
var common_1 = require("../../../models/common");
var id = '4b081d24e2a9430dbc0e3a97d85087a6-948';
exports.POSTAL_TICKET_SETSTATE_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.PRIZE__POSTALTICKET__SETSTATE,
    data: {
        id: id,
        state: PostalTicket_1.PostalTicketStates.PENDING,
    },
};
exports.POSTAL_TICKET_SETSTATE_VALUES_INVALID = {
    type: 'some-type',
    data: {
        id: null,
        state: 'some-state',
    },
};
exports.POSTAL_TICKET_SETSTATE_VALUES = {
    valid: exports.POSTAL_TICKET_SETSTATE_VALUES_VALID,
    invalid: exports.POSTAL_TICKET_SETSTATE_VALUES_INVALID,
};
exports.POSTAL_TICKET_SETSTATES_VALID = {
    DEFAULT: exports.POSTAL_TICKET_SETSTATE_VALUES.valid,
    PENDING: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES.valid, { data: { id: id, state: PostalTicket_1.PostalTicketStates.PENDING } }),
    PROCESSED: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES.valid, { data: { id: id, state: PostalTicket_1.PostalTicketStates.PROCESSED } }),
    REJECTED: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES.valid, { data: { id: id, state: PostalTicket_1.PostalTicketStates.REJECTED } }),
};
exports.POSTAL_TICKET_SETSTATES_INVALID = {
    DEFAULT: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES_VALID, { data: exports.POSTAL_TICKET_SETSTATE_VALUES.invalid.data }),
    INVALID_TYPE: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES_VALID, { type: exports.POSTAL_TICKET_SETSTATE_VALUES.invalid.type }),
    INVALID_DATA: utils_1.override(exports.POSTAL_TICKET_SETSTATE_VALUES_VALID, { data: null }),
};
exports.POSTAL_TICKET_SETSTATES = {
    invalid: exports.POSTAL_TICKET_SETSTATES_INVALID,
    valid: exports.POSTAL_TICKET_SETSTATES_VALID,
};
/**
 * Generates a random PostalTicketSetState
 *
 * @param overrides
 */
var createPostalTicketSetState = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.PRIZE__POSTALTICKET__SETSTATE, data: {
            id: utils_1.getString(10) + "-" + utils_1.getInteger(1, 9999),
            state: utils_1.getRandomElement(Object.values(PostalTicket_1.PostalTicketStates)),
        } }, overrides));
};
exports.createPostalTicketSetState = createPostalTicketSetState;
