// @comicrelief/data-models/schemas/erp/mocks/actions/prize/PrizePut.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prizePutGenerator = exports.PRIZE_PUTS = exports.PRIZE_PUTS_INVALID = exports.PRIZE_PUTS_VALID = exports.PRIZE_PUT_VALUES = exports.PRIZE_PUT_VALUES_INVALID = exports.PRIZE_PUT_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
var common_1 = require("../../../models/common");
var Prize_mock_1 = require("../../doctypes/prize/Prize.mock");
exports.PRIZE_PUT_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.PRIZE__PRIZE__PUT,
    data: Prize_mock_1.PRIZES.valid.DEFAULT,
};
exports.PRIZE_PUT_VALUES_INVALID = {
    type: 'some-type',
    data: Prize_mock_1.PRIZES.invalid.DEFAULT,
};
exports.PRIZE_PUT_VALUES = {
    valid: exports.PRIZE_PUT_VALUES_VALID,
    invalid: exports.PRIZE_PUT_VALUES_INVALID,
};
exports.PRIZE_PUTS_VALID = __assign({ DEFAULT: exports.PRIZE_PUT_VALUES.valid }, utils_1.overrideFromMocks(exports.PRIZE_PUT_VALUES_VALID, 'data', Prize_mock_1.PRIZES.valid));
exports.PRIZE_PUTS_INVALID = __assign({ DEFAULT: utils_1.override(exports.PRIZE_PUT_VALUES_VALID, { data: exports.PRIZE_PUT_VALUES_INVALID.data }), INVALID_TYPE: utils_1.override(exports.PRIZE_PUT_VALUES_VALID, { type: exports.PRIZE_PUT_VALUES.invalid.type }), INVALID_DATA: utils_1.override(exports.PRIZE_PUT_VALUES_VALID, { data: null }), MISSING_TYPE: utils_1.omit(exports.PRIZE_PUT_VALUES.valid, ['type']), MISSING_DATA: utils_1.omit(exports.PRIZE_PUT_VALUES.valid, ['data']) }, utils_1.overrideFromMocks(exports.PRIZE_PUT_VALUES_VALID, 'data', Prize_mock_1.PRIZES.invalid));
exports.PRIZE_PUTS = {
    invalid: exports.PRIZE_PUTS_INVALID,
    valid: exports.PRIZE_PUTS_VALID,
};
/**
 * Generates a random PrizeCreate
 *
 * @param overrides
 */
var prizePutGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.PRIZE__PRIZE__PUT, data: Prize_mock_1.prizeGenerator() }, overrides));
};
exports.prizePutGenerator = prizePutGenerator;
