// @comicrelief/data-models/schemas/erp/mocks/actions/prize/TicketPut.mock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ticketPutGenerator = exports.TICKET_PUTS = exports.TICKET_PUTS_INVALID = exports.TICKET_PUTS_VALID = exports.TICKET_PUT_VALUES = exports.TICKET_PUT_VALUES_INVALID = exports.TICKET_PUT_VALUES_VALID = void 0;
var utils_1 = require("../../../../../utils");
var common_1 = require("../../../models/common");
var Ticket_mock_1 = require("../../doctypes/prize/Ticket.mock");
exports.TICKET_PUT_VALUES_VALID = {
    type: common_1.ErpPayloadTypes.PRIZE__TICKET__PUT,
    data: Ticket_mock_1.TICKETS.valid.DEFAULT,
};
exports.TICKET_PUT_VALUES_INVALID = {
    type: 'some-type',
    data: Ticket_mock_1.TICKETS.invalid.DEFAULT,
};
exports.TICKET_PUT_VALUES = {
    valid: exports.TICKET_PUT_VALUES_VALID,
    invalid: exports.TICKET_PUT_VALUES_INVALID,
};
exports.TICKET_PUTS_VALID = __assign({ DEFAULT: exports.TICKET_PUT_VALUES.valid }, utils_1.overrideFromMocks(exports.TICKET_PUT_VALUES.valid, 'data', Ticket_mock_1.TICKETS.valid));
exports.TICKET_PUTS_INVALID = __assign({ DEFAULT: utils_1.override(exports.TICKET_PUT_VALUES_VALID, { data: exports.TICKET_PUT_VALUES.invalid.data }), INVALID_TYPE: utils_1.override(exports.TICKET_PUT_VALUES_VALID, { type: exports.TICKET_PUT_VALUES.invalid.type }), INVALID_DATA: utils_1.override(exports.TICKET_PUT_VALUES_VALID, { data: null }), MISSING_TYPE: utils_1.omit(exports.TICKET_PUT_VALUES_VALID, ['type']), MISSING_DATA: utils_1.omit(exports.TICKET_PUT_VALUES_VALID, ['data']) }, utils_1.overrideFromMocks(exports.TICKET_PUT_VALUES.valid, 'data', Ticket_mock_1.TICKETS.invalid));
exports.TICKET_PUTS = {
    invalid: exports.TICKET_PUTS_INVALID,
    valid: exports.TICKET_PUTS_VALID,
};
/**
 * Generates a random TicketCreate
 *
 * @param overrides
 */
var ticketPutGenerator = function (overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ type: common_1.ErpPayloadTypes.PRIZE__TICKET__PUT, data: Ticket_mock_1.ticketGenerator() }, overrides));
};
exports.ticketPutGenerator = ticketPutGenerator;
